// botframework-webchat-component/lib/hooks/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "useDictateAbortable", {
  enumerable: true,
  get: function get() {
    return _useDictateAbortable.default;
  }
});
Object.defineProperty(exports, "useFocus", {
  enumerable: true,
  get: function get() {
    return _useFocus.default;
  }
});
Object.defineProperty(exports, "useMakeThumbnail", {
  enumerable: true,
  get: function get() {
    return _useMakeThumbnail.default;
  }
});
Object.defineProperty(exports, "useMicrophoneButtonClick", {
  enumerable: true,
  get: function get() {
    return _MicrophoneButton.useMicrophoneButtonClick;
  }
});
Object.defineProperty(exports, "useMicrophoneButtonDisabled", {
  enumerable: true,
  get: function get() {
    return _MicrophoneButton.useMicrophoneButtonDisabled;
  }
});
Object.defineProperty(exports, "useObserveScrollPosition", {
  enumerable: true,
  get: function get() {
    return _useObserveScrollPosition.default;
  }
});
Object.defineProperty(exports, "useObserveTranscriptFocus", {
  enumerable: true,
  get: function get() {
    return _useObserveTranscriptFocus.default;
  }
});
Object.defineProperty(exports, "useRegisterFocusSendBox", {
  enumerable: true,
  get: function get() {
    return _sendBoxFocus.useRegisterFocusSendBox;
  }
});
Object.defineProperty(exports, "useRenderMarkdownAsHTML", {
  enumerable: true,
  get: function get() {
    return _useRenderMarkdownAsHTML.default;
  }
});
Object.defineProperty(exports, "useScrollDown", {
  enumerable: true,
  get: function get() {
    return _useScrollDown.default;
  }
});
Object.defineProperty(exports, "useScrollTo", {
  enumerable: true,
  get: function get() {
    return _useScrollTo.default;
  }
});
Object.defineProperty(exports, "useScrollToEnd", {
  enumerable: true,
  get: function get() {
    return _useScrollToEnd.default;
  }
});
Object.defineProperty(exports, "useScrollUp", {
  enumerable: true,
  get: function get() {
    return _useScrollUp.default;
  }
});
Object.defineProperty(exports, "useSendBoxSpeechInterimsVisible", {
  enumerable: true,
  get: function get() {
    return _BasicSendBox.useSendBoxSpeechInterimsVisible;
  }
});
Object.defineProperty(exports, "useSendFiles", {
  enumerable: true,
  get: function get() {
    return _useSendFiles.default;
  }
});
Object.defineProperty(exports, "useSendMessage", {
  enumerable: true,
  get: function get() {
    return _useSendMessage.default;
  }
});
Object.defineProperty(exports, "useStyleSet", {
  enumerable: true,
  get: function get() {
    return _useStyleSet.default;
  }
});
Object.defineProperty(exports, "useTextBoxSubmit", {
  enumerable: true,
  get: function get() {
    return _TextBox.useTextBoxSubmit;
  }
});
Object.defineProperty(exports, "useTextBoxValue", {
  enumerable: true,
  get: function get() {
    return _TextBox.useTextBoxValue;
  }
});
Object.defineProperty(exports, "useTypingIndicatorVisible", {
  enumerable: true,
  get: function get() {
    return _BasicTypingIndicator.useTypingIndicatorVisible;
  }
});
Object.defineProperty(exports, "useWebSpeechPonyfill", {
  enumerable: true,
  get: function get() {
    return _useWebSpeechPonyfill.default;
  }
});
var _useDictateAbortable = _interopRequireDefault(require("./useDictateAbortable"));
var _useFocus = _interopRequireDefault(require("./useFocus"));
var _useMakeThumbnail = _interopRequireDefault(require("./useMakeThumbnail"));
var _useObserveScrollPosition = _interopRequireDefault(require("./useObserveScrollPosition"));
var _useObserveTranscriptFocus = _interopRequireDefault(require("./useObserveTranscriptFocus"));
var _useRenderMarkdownAsHTML = _interopRequireDefault(require("./useRenderMarkdownAsHTML"));
var _useScrollDown = _interopRequireDefault(require("./useScrollDown"));
var _useScrollTo = _interopRequireDefault(require("./useScrollTo"));
var _useScrollToEnd = _interopRequireDefault(require("./useScrollToEnd"));
var _useScrollUp = _interopRequireDefault(require("./useScrollUp"));
var _useSendFiles = _interopRequireDefault(require("./useSendFiles"));
var _useSendMessage = _interopRequireDefault(require("./useSendMessage"));
var _useStyleSet = _interopRequireDefault(require("./useStyleSet"));
var _useWebSpeechPonyfill = _interopRequireDefault(require("./useWebSpeechPonyfill"));
var _BasicTypingIndicator = require("../BasicTypingIndicator");
var _BasicSendBox = require("../SendBox/BasicSendBox");
var _MicrophoneButton = require("../SendBox/MicrophoneButton");
var _TextBox = require("../SendBox/TextBox");
var _sendBoxFocus = require("./sendBoxFocus");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
