// botframework-webchat-component/lib/Middleware/Activity/createCoreMiddleware.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = createCoreMiddleware;
var _react = _interopRequireDefault(require("react"));
var _CarouselLayout = _interopRequireDefault(require("../../Activity/CarouselLayout"));
var _StackedLayout = _interopRequireDefault(require("../../Activity/StackedLayout"));
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function _extends() { _extends = Object.assign ? Object.assign.bind() : function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }
function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }
function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }
function createCoreMiddleware() {
  return [function () {
    return function (next) {
      return function () {
        for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
          args[_key] = arguments[_key];
        }
        var activity = args[0].activity;

        // TODO: [P4] Can we simplify these if-statement to something more readable?

        var type = activity.type;
        if (type === 'typing') {
          if (!('text' in activity && typeof activity.text === 'string' && activity.channelData.streamType !== 'informative')) {
            // If it is an informative message, hide it until we have a design for informative message.
            return false;
          }

          // Should show if this is useActiveTyping()[0][*].firstActivity, and render it with the content of lastActivity.
          return function renderStackedLayout(renderAttachment, props) {
            typeof props === 'undefined' && console.warn('botframework-webchat: One or more arguments were missing after passing through the activity middleware. Please check your custom activity middleware to make sure it passes all arguments.');
            return /*#__PURE__*/_react.default.createElement(_StackedLayout.default, _extends({
              activity: _objectSpread(_objectSpread({}, activity), {}, {
                type: 'message'
              }),
              renderAttachment: renderAttachment
            }, props));
          };
        }

        // Filter out activities that should not be visible
        if (type === 'conversationUpdate' || type === 'event' || type === 'invoke') {
          return false;
        } else if (type === 'message') {
          var attachments = activity.attachments,
            channelData = activity.channelData,
            text = activity.text;
          if (
          // Do not show postback
          channelData !== null && channelData !== void 0 && channelData.postBack ||
          // Do not show messageBack if displayText is undefined
          channelData !== null && channelData !== void 0 && channelData.messageBack && !channelData.messageBack.displayText ||
          // Do not show empty bubbles (no text and attachments, and not "typing")
          !(text || attachments !== null && attachments !== void 0 && attachments.length)) {
            return false;
          }
        }
        if (type === 'message' || type === 'typing') {
          var _activity$attachments;
          if (type === 'message' && (((_activity$attachments = activity.attachments) === null || _activity$attachments === void 0 ? void 0 : _activity$attachments.length) || 0) > 1 && activity.attachmentLayout === 'carousel') {
            // The following line is not a React functional component, it's a render function called by useCreateActivityRenderer() hook.
            // The function signature need to be compatible with older version of activity middleware, which was:
            //
            // renderActivity(
            //   renderAttachment: ({ activity, attachment }) => React.Element
            // ) => React.Element

            return function renderCarouselLayout(renderAttachment, props) {
              typeof props === 'undefined' && console.warn('botframework-webchat: One or more arguments were missing after passing through the activity middleware. Please check your custom activity middleware to make sure it passes all arguments.');
              return /*#__PURE__*/_react.default.createElement(_CarouselLayout.default, _extends({
                activity: activity,
                renderAttachment: renderAttachment
              }, props));
            };
          }

          // The following line is not a React functional component, it's a render function called by useCreateActivityRenderer() hook.
          return function renderStackedLayout(renderAttachment, props) {
            typeof props === 'undefined' && console.warn('botframework-webchat: One or more arguments were missing after passing through the activity middleware. Please check your custom activity middleware to make sure it passes all arguments.');
            return /*#__PURE__*/_react.default.createElement(_StackedLayout.default, _extends({
              activity: activity,
              renderAttachment: renderAttachment
            }, props));
          };
        }
        return next.apply(void 0, args);
      };
    };
  }];
}
