// botframework-webchat-component/lib/Attachment/Text/private/MarkdownTextContent.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (obj) { return typeof obj; } : function (obj) { return obj && "function" == typeof Symbol && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }, _typeof(obj); }
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _botframeworkWebchatApi = require("botframework-webchat-api");
var _botframeworkWebchatCore = require("botframework-webchat-core");
var _classnames = _interopRequireDefault(require("classnames"));
var _mdastUtilFromMarkdown = require("mdast-util-from-markdown");
var _react = _interopRequireWildcard(require("react"));
var _useRefFrom = require("use-ref-from");
var _index = require("../../../LinkDefinition/index");
var _dereferenceBlankNodes = _interopRequireDefault(require("../../../Utils/JSONLinkedData/dereferenceBlankNodes"));
var _useRenderMarkdownAsHTML = _interopRequireDefault(require("../../../hooks/useRenderMarkdownAsHTML"));
var _useStyleSet3 = _interopRequireDefault(require("../../../hooks/useStyleSet"));
var _useShowModal = _interopRequireDefault(require("../../../providers/ModalDialog/useShowModal"));
var _CitationModalContent = _interopRequireDefault(require("./CitationModalContent"));
var _MessageSensitivityLabel = _interopRequireDefault(require("./MessageSensitivityLabel"));
var _isHTMLButtonElement = _interopRequireDefault(require("./isHTMLButtonElement"));
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }
function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }
function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }
function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }
function _iterableToArrayLimit(arr, i) { var _i = arr == null ? null : typeof Symbol !== "undefined" && arr[Symbol.iterator] || arr["@@iterator"]; if (_i == null) return; var _arr = []; var _n = true; var _d = false; var _s, _e; try { for (_i = _i.call(arr); !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }
function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }
var useLocalizer = _botframeworkWebchatApi.hooks.useLocalizer;
function isCitationURL(url) {
  var _onErrorResumeNext;
  return ((_onErrorResumeNext = (0, _botframeworkWebchatCore.onErrorResumeNext)(function () {
    return new URL(url);
  })) === null || _onErrorResumeNext === void 0 ? void 0 : _onErrorResumeNext.protocol) === 'cite:';
}
var MarkdownTextContent = /*#__PURE__*/(0, _react.memo)(function (_ref) {
  var activity = _ref.activity,
    markdown = _ref.markdown;
  var _useStyleSet = (0, _useStyleSet3.default)(),
    _useStyleSet2 = _slicedToArray(_useStyleSet, 1),
    _useStyleSet2$ = _useStyleSet2[0],
    citationModalDialogStyleSet = _useStyleSet2$.citationModalDialog,
    renderMarkdownStyleSet = _useStyleSet2$.renderMarkdown,
    textContentStyleSet = _useStyleSet2$.textContent;
  var localize = useLocalizer();
  var graph = (0, _react.useMemo)(function () {
    return (0, _dereferenceBlankNodes.default)(activity.entities || []);
  }, [activity.entities]);
  var renderMarkdownAsHTML = (0, _useRenderMarkdownAsHTML.default)('message activity');
  var showModal = (0, _useShowModal.default)();
  var messageThing = (0, _react.useMemo)(function () {
    return (0, _botframeworkWebchatCore.getOrgSchemaMessage)(graph);
  }, [graph]);
  var citationModalDialogLabel = localize('CITATION_MODEL_DIALOG_ALT');
  if (!renderMarkdownAsHTML) {
    throw new Error('botframework-webchat: assert failed for renderMarkdownAsHTML');
  }
  var dangerouslySetInnerHTML = (0, _react.useMemo)(function () {
    return {
      __html: markdown ? renderMarkdownAsHTML(markdown) : ''
    };
  }, [renderMarkdownAsHTML, markdown]);
  var markdownDefinitions = (0, _react.useMemo)(function () {
    return (0, _mdastUtilFromMarkdown.fromMarkdown)(markdown).children.filter(function (node) {
      return node.type === 'definition';
    });
  }, [markdown]);
  var showClaimModal = (0, _react.useCallback)(function (title, text, altText) {
    showModal(function () {
      return /*#__PURE__*/_react.default.createElement(_CitationModalContent.default, {
        headerText: title,
        markdown: text
      });
    }, {
      'aria-label': altText || title || citationModalDialogLabel,
      className: (0, _classnames.default)('webchat__citation-modal-dialog', citationModalDialogStyleSet)
    });
  }, [citationModalDialogStyleSet, citationModalDialogLabel, showModal]);
  var entries = (0, _react.useMemo)(function () {
    return Object.freeze(markdownDefinitions.map(function (markdownDefinition) {
      var _messageThing$citatio;
      var messageCitation = messageThing === null || messageThing === void 0 ? void 0 : (_messageThing$citatio = messageThing.citation) === null || _messageThing$citatio === void 0 ? void 0 : _messageThing$citatio.map(_botframeworkWebchatCore.parseClaim).find(function (_ref2) {
        var position = _ref2.position;
        return '' + position === markdownDefinition.identifier;
      });
      if (messageCitation) {
        return {
          claim: messageCitation,
          key: markdownDefinition.url,
          handleClick: messageCitation !== null && messageCitation !== void 0 && messageCitation.appearance && !messageCitation.appearance.url ? function () {
            var _messageCitation$appe;
            return showClaimModal((_messageCitation$appe = messageCitation.appearance.name) !== null && _messageCitation$appe !== void 0 ? _messageCitation$appe : markdownDefinition.title, messageCitation.appearance.text, messageCitation.alternateName);
          } : undefined,
          markdownDefinition: markdownDefinition,
          url: messageCitation !== null && messageCitation !== void 0 && messageCitation.appearance ? messageCitation.appearance.url : markdownDefinition.url
        };
      }
      var rootLevelClaim = graph.filter(function (_ref3) {
        var type = _ref3.type;
        return type === 'https://schema.org/Claim';
      }).map(_botframeworkWebchatCore.parseClaim).find(function (_ref4) {
        var id = _ref4['@id'];
        return id === markdownDefinition.url;
      });
      if (rootLevelClaim) {
        return {
          claim: rootLevelClaim,
          key: markdownDefinition.url,
          handleClick: isCitationURL(rootLevelClaim['@id']) ? function () {
            var _rootLevelClaim$name;
            return showClaimModal((_rootLevelClaim$name = rootLevelClaim.name) !== null && _rootLevelClaim$name !== void 0 ? _rootLevelClaim$name : markdownDefinition.title, rootLevelClaim.text, rootLevelClaim.alternateName);
          } : undefined,
          markdownDefinition: markdownDefinition
        };
      }
      return {
        key: markdownDefinition.url,
        markdownDefinition: markdownDefinition,
        url: markdownDefinition.url
      };
    }));
  }, [graph, markdownDefinitions, messageThing, showClaimModal]);
  var entriesRef = (0, _useRefFrom.useRefFrom)(entries);
  var handleClick = (0, _react.useCallback)(function (event) {
    // Find out what <button> is being clicked.
    var targetElement = event.target;
    var buttonElement = (0, _isHTMLButtonElement.default)(targetElement) ? targetElement : targetElement.closest('button');
    if (!buttonElement || !targetElement.contains(buttonElement)) {
      return;
    }
    var entry = entriesRef.current.find(function (_ref5) {
      var key = _ref5.key;
      return key === buttonElement.value;
    });
    if (entry !== null && entry !== void 0 && entry.handleClick) {
      event.preventDefault();
      event.stopPropagation();
      entry.handleClick();
    }
  }, [entriesRef]);
  var messageSensitivityLabelProps = (0, _react.useMemo)(function () {
    var usageInfo = messageThing === null || messageThing === void 0 ? void 0 : messageThing.usageInfo;
    if (usageInfo) {
      var _usageInfo$keywords;
      var pattern = usageInfo.pattern;
      var encryptionStatus = !!((_usageInfo$keywords = usageInfo.keywords) !== null && _usageInfo$keywords !== void 0 && _usageInfo$keywords.find(function (keyword) {
        return keyword === 'encrypted-content';
      }));
      return {
        color: pattern && pattern.inDefinedTermSet === 'https://www.w3.org/TR/css-color-4/' && pattern.name === 'color' && pattern.termCode,
        isEncrypted: encryptionStatus,
        name: usageInfo.name,
        title: usageInfo.description
      };
    }
  }, [messageThing]);

  // The main text of the citation entry (e.g. the title of the document). Used as the content of the main link and, if it exists, the header of the popup window.
  var getEntryMainText = function getEntryMainText(entry) {
    var _ref6, _entry$claim$name, _entry$claim, _entry$claim2, _entry$claim2$appeara;
    return (_ref6 = (_entry$claim$name = (_entry$claim = entry.claim) === null || _entry$claim === void 0 ? void 0 : _entry$claim.name) !== null && _entry$claim$name !== void 0 ? _entry$claim$name : (_entry$claim2 = entry.claim) === null || _entry$claim2 === void 0 ? void 0 : (_entry$claim2$appeara = _entry$claim2.appearance) === null || _entry$claim2$appeara === void 0 ? void 0 : _entry$claim2$appeara.name) !== null && _ref6 !== void 0 ? _ref6 : entry.markdownDefinition.title;
  };

  // Optional alternate name for the entry, used as a subtitle beneath the link
  var getEntryBadgeName = function getEntryBadgeName(entry) {
    var _entry$claim3, _entry$claim3$appeara, _entry$claim3$appeara2;
    return (_entry$claim3 = entry.claim) === null || _entry$claim3 === void 0 ? void 0 : (_entry$claim3$appeara = _entry$claim3.appearance) === null || _entry$claim3$appeara === void 0 ? void 0 : (_entry$claim3$appeara2 = _entry$claim3$appeara.usageInfo) === null || _entry$claim3$appeara2 === void 0 ? void 0 : _entry$claim3$appeara2.name;
  };

  // Secondary text describing the citation, used in the a11y description (i.e. the div's title attribute)
  var getEntryDescription = function getEntryDescription(entry) {
    var _entry$claim4, _entry$claim4$appeara, _entry$claim4$appeara2;
    return (_entry$claim4 = entry.claim) === null || _entry$claim4 === void 0 ? void 0 : (_entry$claim4$appeara = _entry$claim4.appearance) === null || _entry$claim4$appeara === void 0 ? void 0 : (_entry$claim4$appeara2 = _entry$claim4$appeara.usageInfo) === null || _entry$claim4$appeara2 === void 0 ? void 0 : _entry$claim4$appeara2.description;
  };
  return /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _classnames.default)('webchat__text-content', 'webchat__text-content--is-markdown', textContentStyleSet + '')
  }, /*#__PURE__*/_react.default.createElement("div", {
    className: (0, _classnames.default)('webchat__text-content__markdown', renderMarkdownStyleSet + '')
    // eslint-disable-next-line react/no-danger
    ,
    dangerouslySetInnerHTML: dangerouslySetInnerHTML,
    onClick: handleClick
  }), !!entries.length && /*#__PURE__*/_react.default.createElement(_index.LinkDefinitions, {
    accessoryComponentType: messageSensitivityLabelProps && _MessageSensitivityLabel.default,
    accessoryProps: messageSensitivityLabelProps
  }, entries.map(function (entry) {
    var _getEntryBadgeName, _getEntryDescription;
    return /*#__PURE__*/_react.default.createElement(_index.LinkDefinitionItem, {
      badgeName: getEntryBadgeName(entry),
      badgeTitle: "".concat((_getEntryBadgeName = getEntryBadgeName(entry)) !== null && _getEntryBadgeName !== void 0 ? _getEntryBadgeName : '', "\n\n").concat((_getEntryDescription = getEntryDescription(entry)) !== null && _getEntryDescription !== void 0 ? _getEntryDescription : '').trim(),
      identifier: entry.markdownDefinition.label,
      key: entry.key,
      onClick: entry.handleClick,
      text: getEntryMainText(entry),
      url: entry.url
    });
  })));
});
MarkdownTextContent.displayName = 'MarkdownTextContent';
var _default = MarkdownTextContent;
exports.default = _default;
