// botframework-webchat/lib/AddFullBundle.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _botframeworkWebchatCore = require("botframework-webchat-core");
var _react = _interopRequireDefault(require("react"));
var _AdaptiveCardsComposer = _interopRequireDefault(require("./adaptiveCards/AdaptiveCardsComposer"));
var _useComposerProps = _interopRequireDefault(require("./useComposerProps"));
var adaptiveCardHostConfigDeprecation = (0, _botframeworkWebchatCore.warnOnce)('"adaptiveCardHostConfig" is deprecated. Please use "adaptiveCardsHostConfig" instead. "adaptiveCardHostConfig" will be removed on or after 2022-01-01.');
var AddFullBundle = function AddFullBundle(_ref) {
  var adaptiveCardHostConfig = _ref.adaptiveCardHostConfig,
    adaptiveCardsHostConfig = _ref.adaptiveCardsHostConfig,
    adaptiveCardsPackage = _ref.adaptiveCardsPackage,
    attachmentForScreenReaderMiddleware = _ref.attachmentForScreenReaderMiddleware,
    attachmentMiddleware = _ref.attachmentMiddleware,
    children = _ref.children,
    renderMarkdown = _ref.renderMarkdown,
    styleOptions = _ref.styleOptions,
    styleSet = _ref.styleSet;
  adaptiveCardHostConfig && adaptiveCardHostConfigDeprecation();
  var patchedProps = (0, _useComposerProps.default)({
    attachmentForScreenReaderMiddleware: (0, _botframeworkWebchatCore.singleToArray)(attachmentForScreenReaderMiddleware),
    attachmentMiddleware: (0, _botframeworkWebchatCore.singleToArray)(attachmentMiddleware),
    renderMarkdown: renderMarkdown,
    styleOptions: styleOptions,
    styleSet: styleSet
  });
  return /*#__PURE__*/_react.default.createElement(_AdaptiveCardsComposer.default, {
    adaptiveCardsHostConfig: adaptiveCardHostConfig || adaptiveCardsHostConfig,
    adaptiveCardsPackage: adaptiveCardsPackage
  }, children(patchedProps));
};
var _default = AddFullBundle;
exports.default = _default;
