// botframework-webchat/lib/adaptiveCards/Attachment/AnimationCardContent.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));
var _botframeworkWebchatComponent = require("botframework-webchat-component");
var _propTypes = _interopRequireDefault(require("prop-types"));
var _react = _interopRequireDefault(require("react"));
var _CommonCard = _interopRequireDefault(require("./CommonCard"));
var _useStyleSet3 = _interopRequireDefault(require("../../hooks/useStyleSet"));
/* eslint react/no-array-index-key: "off" */

var ImageContent = _botframeworkWebchatComponent.Components.ImageContent,
  VideoContent = _botframeworkWebchatComponent.Components.VideoContent;
var AnimationCardContent = function AnimationCardContent(_ref) {
  var actionPerformedClassName = _ref.actionPerformedClassName,
    content = _ref.content,
    disabled = _ref.disabled;
  var _content$media = content.media,
    media = _content$media === void 0 ? [] : _content$media;
  var _useStyleSet = (0, _useStyleSet3.default)(),
    _useStyleSet2 = (0, _slicedToArray2.default)(_useStyleSet, 1),
    animationCardAttachmentStyleSet = _useStyleSet2[0].animationCardAttachment;
  return /*#__PURE__*/_react.default.createElement("div", {
    className: animationCardAttachmentStyleSet
  }, /*#__PURE__*/_react.default.createElement("ul", {
    className: "media-list"
  }, media.map(function (_ref2, index) {
    var _ref2$profile = _ref2.profile,
      profile = _ref2$profile === void 0 ? '' : _ref2$profile,
      url = _ref2.url;
    return /*#__PURE__*/_react.default.createElement("li", {
      key: index
    }, /\.gif$/i.test(url) ? /*#__PURE__*/_react.default.createElement(ImageContent, {
      alt: profile,
      src: url
    }) : /*#__PURE__*/_react.default.createElement(VideoContent, {
      alt: profile,
      src: url
    }));
  })), /*#__PURE__*/_react.default.createElement(_CommonCard.default, {
    actionPerformedClassName: actionPerformedClassName,
    content: content,
    disabled: disabled
  }));
};
AnimationCardContent.defaultProps = {
  actionPerformedClassName: '',
  disabled: undefined
};
AnimationCardContent.propTypes = {
  actionPerformedClassName: _propTypes.default.string,
  // PropTypes cannot fully capture TypeScript types.
  // @ts-ignore
  content: _propTypes.default.shape({
    media: _propTypes.default.arrayOf(_propTypes.default.shape({
      profile: _propTypes.default.string,
      url: _propTypes.default.string.isRequired
    })).isRequired
  }).isRequired,
  disabled: _propTypes.default.bool
};
var _default = AnimationCardContent;
exports.default = _default;
