// botframework-webchat/lib/adaptiveCards/Attachment/VideoCardContent.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));
var _botframeworkWebchatComponent = require("botframework-webchat-component");
var _propTypes = _interopRequireDefault(require("prop-types"));
var _react = _interopRequireDefault(require("react"));
var _CommonCard = _interopRequireDefault(require("./CommonCard"));
var _useStyleSet3 = _interopRequireDefault(require("../../hooks/useStyleSet"));
/* eslint react/no-array-index-key: "off" */

var VideoContent = _botframeworkWebchatComponent.Components.VideoContent;
var VideoCardContent = function VideoCardContent(_ref) {
  var actionPerformedClassName = _ref.actionPerformedClassName,
    content = _ref.content,
    disabled = _ref.disabled;
  var autoloop = content.autoloop,
    autostart = content.autostart,
    _content$image = content.image;
  _content$image = _content$image === void 0 ? {
    url: undefined
  } : _content$image;
  var imageURL = _content$image.url,
    media = content.media;
  var _useStyleSet = (0, _useStyleSet3.default)(),
    _useStyleSet2 = (0, _slicedToArray2.default)(_useStyleSet, 1),
    audioCardAttachmentStyleSet = _useStyleSet2[0].audioCardAttachment;
  return /*#__PURE__*/_react.default.createElement("div", {
    className: audioCardAttachmentStyleSet
  }, /*#__PURE__*/_react.default.createElement("ul", {
    className: "media-list"
  }, media.map(function (_ref2, index) {
    var url = _ref2.url;
    return /*#__PURE__*/_react.default.createElement("li", {
      key: index
    }, /*#__PURE__*/_react.default.createElement(VideoContent, {
      autoPlay: autostart,
      loop: autoloop,
      poster: imageURL,
      src: url
    }));
  })), /*#__PURE__*/_react.default.createElement(_CommonCard.default, {
    actionPerformedClassName: actionPerformedClassName,
    content: content,
    disabled: disabled
  }));
};
VideoCardContent.defaultProps = {
  actionPerformedClassName: '',
  disabled: undefined
};
VideoCardContent.propTypes = {
  actionPerformedClassName: _propTypes.default.string,
  // PropTypes cannot fully capture TypeScript types.
  // @ts-ignore
  content: _propTypes.default.shape({
    autoloop: _propTypes.default.bool,
    autostart: _propTypes.default.bool,
    image: _propTypes.default.shape({
      url: _propTypes.default.string.isRequired
    }),
    media: _propTypes.default.arrayOf(_propTypes.default.shape({
      profile: _propTypes.default.string,
      url: _propTypes.default.string.isRequired
    })).isRequired
  }).isRequired,
  disabled: _propTypes.default.bool
};
var _default = VideoCardContent;
exports.default = _default;
