// botframework-webchat-component/lib/TranscriptActivity.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (obj) { return typeof obj; } : function (obj) { return obj && "function" == typeof Symbol && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }, _typeof(obj); }
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireWildcard(require("react"));
var _botframeworkWebchatApi = require("botframework-webchat-api");
var _ActivityRow = _interopRequireDefault(require("./Transcript/ActivityRow"));
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
var useCreateActivityStatusRenderer = _botframeworkWebchatApi.hooks.useCreateActivityStatusRenderer;
function TranscriptActivity(_ref) {
  var activityElementMapRef = _ref.activityElementMapRef,
    activityKey = _ref.activityKey,
    activity = _ref.activity,
    hideTimestamp = _ref.hideTimestamp,
    renderActivity = _ref.renderActivity,
    renderAvatar = _ref.renderAvatar,
    showCallout = _ref.showCallout;
  var createActivityStatusRenderer = useCreateActivityStatusRenderer();
  var activityCallbackRef = (0, _react.useCallback)(function (activityElement) {
    activityElement ? activityElementMapRef.current.set(activityKey, activityElement) : activityElementMapRef.current.delete(activityKey);
  }, [activityElementMapRef, activityKey]);
  var renderActivityStatus = (0, _react.useMemo)(function () {
    return createActivityStatusRenderer({
      activity: activity,
      nextVisibleActivity: undefined
    });
  }, [activity, createActivityStatusRenderer]);
  var children = (0, _react.useMemo)(function () {
    return renderActivity({
      hideTimestamp: hideTimestamp,
      renderActivityStatus: renderActivityStatus,
      renderAvatar: renderAvatar,
      showCallout: showCallout
    });
  }, [hideTimestamp, renderActivity, renderActivityStatus, renderAvatar, showCallout]);
  return /*#__PURE__*/_react.default.createElement(_ActivityRow.default, {
    activity: activity,
    ref: activityCallbackRef
  }, children);
}
var _default = /*#__PURE__*/(0, _react.memo)(TranscriptActivity);
exports.default = _default;
