// use-propagate/lib/commonjs/createPropagation.js
"use strict";

var _Object$defineProperty = require("@babel/runtime-corejs3/core-js-stable/object/define-property");
var _interopRequireDefault = require("@babel/runtime-corejs3/helpers/interopRequireDefault").default;
_Object$defineProperty(exports, "__esModule", {
  value: true
});
exports.default = createPropagation;
var _set = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/set"));
var _forEach = _interopRequireDefault(require("@babel/runtime-corejs3/core-js-stable/instance/for-each"));
var _react = require("react");
var _useRefFrom = require("use-ref-from");
function createPropagation() {
  var init = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  var rendering = false;
  var allowPropagateDuringRender = init.allowPropagateDuringRender;
  var listeners = new _set.default();
  var addListener = function addListener(listener) {
    return void listeners.add(listener);
  };
  var removeListener = function removeListener(listener) {
    return void listeners.delete(listener);
  };
  var usePropagate = function usePropagate() {
    rendering = true;
    (0, _react.useLayoutEffect)(function () {
      rendering = false;
    });
    return function (value) {
      if (rendering && !allowPropagateDuringRender) {
        return console.warn('use-propagate: The propagate callback function should not be called while rendering, ignoring the call.');
      }
      (0, _forEach.default)(listeners).call(listeners, function (listener) {
        return listener(value);
      });
    };
  };
  return {
    useListen: function useListen(listener) {
      var listenerRef = (0, _useRefFrom.useRefFrom)(listener);
      var wrappingListener = (0, _react.useCallback)(function (value) {
        return listenerRef.current(value);
      }, [listenerRef]);
      (0, _react.useMemo)(function () {
        return addListener(wrappingListener);
      }, [wrappingListener]);
      (0, _react.useEffect)(function () {
        return function () {
          return removeListener(wrappingListener);
        };
      }, [wrappingListener]);
    },
    usePropagate: usePropagate
  };
}
