// botframework-webchat-component/lib/hooks/internal/BypassSpeechSynthesisPonyfill.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (obj) { return typeof obj; } : function (obj) { return obj && "function" == typeof Symbol && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }, _typeof(obj); }
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.speechSynthesis = exports.SpeechSynthesisVoice = exports.SpeechSynthesisUtterance = exports.SpeechSynthesisEvent = void 0;
var _createOnEventShim = _interopRequireDefault(require("./private/createOnEventShim"));
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }
function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }
function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); Object.defineProperty(Constructor, "prototype", { writable: false }); return Constructor; }
function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); Object.defineProperty(subClass, "prototype", { writable: false }); if (superClass) _setPrototypeOf(subClass, superClass); }
function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }
function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } else if (call !== void 0) { throw new TypeError("Derived constructors may only return object or undefined"); } return _assertThisInitialized(self); }
function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }
function _wrapNativeSuper(Class) { var _cache = typeof Map === "function" ? new Map() : undefined; _wrapNativeSuper = function _wrapNativeSuper(Class) { if (Class === null || !_isNativeFunction(Class)) return Class; if (typeof Class !== "function") { throw new TypeError("Super expression must either be null or a function"); } if (typeof _cache !== "undefined") { if (_cache.has(Class)) return _cache.get(Class); _cache.set(Class, Wrapper); } function Wrapper() { return _construct(Class, arguments, _getPrototypeOf(this).constructor); } Wrapper.prototype = Object.create(Class.prototype, { constructor: { value: Wrapper, enumerable: false, writable: true, configurable: true } }); return _setPrototypeOf(Wrapper, Class); }; return _wrapNativeSuper(Class); }
function _construct(Parent, args, Class) { if (_isNativeReflectConstruct()) { _construct = Reflect.construct.bind(); } else { _construct = function _construct(Parent, args, Class) { var a = [null]; a.push.apply(a, args); var Constructor = Function.bind.apply(Parent, a); var instance = new Constructor(); if (Class) _setPrototypeOf(instance, Class.prototype); return instance; }; } return _construct.apply(null, arguments); }
function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }
function _isNativeFunction(fn) { return Function.toString.call(fn).indexOf("[native code]") !== -1; }
function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf ? Object.setPrototypeOf.bind() : function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }
function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf.bind() : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); } // Since this is a bypass, we will relax some ESLint rules.
// All classes/properties defined here are in W3C Web Speech API.
/* eslint class-methods-use-this: "off" */ /* eslint getter-return: "off" */ /* eslint max-classes-per-file: ["error", 4] */ /* eslint no-empty-function: "off" */
var SpeechSynthesisEvent = /*#__PURE__*/function (_Event) {
  _inherits(SpeechSynthesisEvent, _Event);
  var _super = _createSuper(SpeechSynthesisEvent);
  function SpeechSynthesisEvent(type, utterance) {
    var _this;
    _classCallCheck(this, SpeechSynthesisEvent);
    _this = _super.call(this, type);
    _this._utterance = utterance;
    return _this;
  }
  _createClass(SpeechSynthesisEvent, [{
    key: "charIndex",
    get: function get() {
      return 0;
    }
  }, {
    key: "elapsedTime",
    get: function get() {
      return 0;
    }
  }, {
    key: "name",
    get: function get() {
      // It is expected to return `undefined`, keeping the function empty to reduce footprint.
    }
  }, {
    key: "utterance",
    get: function get() {
      return this._utterance;
    }
  }]);
  return SpeechSynthesisEvent;
}( /*#__PURE__*/_wrapNativeSuper(Event));
exports.SpeechSynthesisEvent = SpeechSynthesisEvent;
var SpeechSynthesisUtterance = /*#__PURE__*/function (_EventTarget) {
  _inherits(SpeechSynthesisUtterance, _EventTarget);
  var _super2 = _createSuper(SpeechSynthesisUtterance);
  function SpeechSynthesisUtterance(text) {
    var _this2;
    _classCallCheck(this, SpeechSynthesisUtterance);
    _this2 = _super2.call(this);
    _this2._lang = 'en-US';
    _this2._pitch = 1;
    _this2._rate = 1;
    _this2._text = text;
    _this2._voice = null;
    _this2._volume = 1;
    _this2._shim = (0, _createOnEventShim.default)(_assertThisInitialized(_this2));
    return _this2;
  }
  _createClass(SpeechSynthesisUtterance, [{
    key: "lang",
    get: function get() {
      return this._lang;
    },
    set: function set(value) {
      this._lang = value;
    }
  }, {
    key: "pitch",
    get: function get() {
      return this._pitch;
    },
    set: function set(value) {
      this._pitch = value;
    }
  }, {
    key: "rate",
    get: function get() {
      return this._rate;
    },
    set: function set(value) {
      this._rate = value;
    }
  }, {
    key: "text",
    get: function get() {
      return this._text;
    },
    set: function set(value) {
      this._text = value;
    }
  }, {
    key: "voice",
    get: function get() {
      return this._voice;
    },
    set: function set(value) {
      this._voice = value;
    }
  }, {
    key: "volume",
    get: function get() {
      return this._volume;
    },
    set: function set(value) {
      this._volume = value;
    }
  }, {
    key: "onboundary",
    get: function get() {
      return this._shim.get('boundary');
    },
    set: function set(value) {
      this._shim.set('boundary', value);
    }
  }, {
    key: "onend",
    get: function get() {
      return this._shim.get('end');
    },
    set: function set(value) {
      this._shim.set('end', value);
    }
  }, {
    key: "onerror",
    get: function get() {
      return this._shim.get('error');
    },
    set: function set(value) {
      this._shim.set('error', value);
    }
  }, {
    key: "onmark",
    get: function get() {
      return this._shim.get('mark');
    },
    set: function set(value) {
      this._shim.set('mark', value);
    }
  }, {
    key: "onpause",
    get: function get() {
      return this._shim.get('pause');
    },
    set: function set(value) {
      this._shim.set('pause', value);
    }
  }, {
    key: "onresume",
    get: function get() {
      return this._shim.get('resume');
    },
    set: function set(value) {
      this._shim.set('resume', value);
    }
  }, {
    key: "onstart",
    get: function get() {
      return this._shim.get('start');
    },
    set: function set(value) {
      this._shim.set('start', value);
    }
  }]);
  return SpeechSynthesisUtterance;
}( /*#__PURE__*/_wrapNativeSuper(EventTarget));
exports.SpeechSynthesisUtterance = SpeechSynthesisUtterance;
var SpeechSynthesisVoice = /*#__PURE__*/function () {
  function SpeechSynthesisVoice() {
    _classCallCheck(this, SpeechSynthesisVoice);
  }
  _createClass(SpeechSynthesisVoice, [{
    key: "default",
    get: function get() {
      return true;
    }
  }, {
    key: "lang",
    get: function get() {
      return 'en-US';
    }
  }, {
    key: "localService",
    get: function get() {
      return true;
    }
  }, {
    key: "name",
    get: function get() {
      return 'English (US)';
    }
  }, {
    key: "voiceURI",
    get: function get() {
      return 'English (US)';
    }
  }]);
  return SpeechSynthesisVoice;
}();
exports.SpeechSynthesisVoice = SpeechSynthesisVoice;
var SpeechSynthesis = /*#__PURE__*/function (_EventTarget2) {
  _inherits(SpeechSynthesis, _EventTarget2);
  var _super3 = _createSuper(SpeechSynthesis);
  function SpeechSynthesis() {
    var _this3;
    _classCallCheck(this, SpeechSynthesis);
    _this3 = _super3.call(this);
    _this3._shim = (0, _createOnEventShim.default)(_assertThisInitialized(_this3));
    return _this3;
  }
  _createClass(SpeechSynthesis, [{
    key: "paused",
    get: function get() {
      return false;
    }
  }, {
    key: "pending",
    get: function get() {
      return false;
    }
  }, {
    key: "speaking",
    get: function get() {
      return false;
    }
  }, {
    key: "cancel",
    value: function cancel() {}
  }, {
    key: "getVoices",
    value: function getVoices() {
      return [new SpeechSynthesisVoice()];
    }
  }, {
    key: "pause",
    value: function pause() {
      throw new Error('pause is not implemented.');
    }
  }, {
    key: "resume",
    value: function resume() {
      throw new Error('resume is not implemented.');
    }
  }, {
    key: "speak",
    value: function speak(utterance) {
      utterance.dispatchEvent(new SpeechSynthesisEvent('start', utterance));
      utterance.dispatchEvent(new SpeechSynthesisEvent('end', utterance));
    }
  }, {
    key: "onvoiceschanged",
    get: function get() {
      return this._shim.get('voiceschanged');
    },
    set: function set(value) {
      this._shim.set('voiceschanged', value);
    }
  }]);
  return SpeechSynthesis;
}( /*#__PURE__*/_wrapNativeSuper(EventTarget));
var speechSynthesis = new SpeechSynthesis();
exports.speechSynthesis = speechSynthesis;
