// botframework-webchat-core/lib/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.buildInfo = exports.Constants = void 0;
Object.defineProperty(exports, "clearSuggestedActions", {
  enumerable: true,
  get: function get() {
    return _clearSuggestedActions["default"];
  }
});
Object.defineProperty(exports, "connect", {
  enumerable: true,
  get: function get() {
    return _connect["default"];
  }
});
Object.defineProperty(exports, "createStore", {
  enumerable: true,
  get: function get() {
    return _createStore["default"];
  }
});
Object.defineProperty(exports, "createStoreWithDevTools", {
  enumerable: true,
  get: function get() {
    return _createStore.withDevTools;
  }
});
Object.defineProperty(exports, "createStoreWithOptions", {
  enumerable: true,
  get: function get() {
    return _createStore.withOptions;
  }
});
Object.defineProperty(exports, "disconnect", {
  enumerable: true,
  get: function get() {
    return _disconnect["default"];
  }
});
Object.defineProperty(exports, "dismissNotification", {
  enumerable: true,
  get: function get() {
    return _dismissNotification["default"];
  }
});
Object.defineProperty(exports, "emitTypingIndicator", {
  enumerable: true,
  get: function get() {
    return _emitTypingIndicator["default"];
  }
});
Object.defineProperty(exports, "getOrgSchemaMessage", {
  enumerable: true,
  get: function get() {
    return _getOrgSchemaMessage["default"];
  }
});
Object.defineProperty(exports, "isForbiddenPropertyName", {
  enumerable: true,
  get: function get() {
    return _isForbiddenPropertyName["default"];
  }
});
Object.defineProperty(exports, "markActivity", {
  enumerable: true,
  get: function get() {
    return _markActivity["default"];
  }
});
exports.moduleFormat = void 0;
Object.defineProperty(exports, "onErrorResumeNext", {
  enumerable: true,
  get: function get() {
    return _onErrorResumeNext["default"];
  }
});
Object.defineProperty(exports, "parseAction", {
  enumerable: true,
  get: function get() {
    return _Action.parseAction;
  }
});
Object.defineProperty(exports, "parseClaim", {
  enumerable: true,
  get: function get() {
    return _Claim.parseClaim;
  }
});
Object.defineProperty(exports, "parseCreativeWork", {
  enumerable: true,
  get: function get() {
    return _CreativeWork.parseCreativeWork;
  }
});
Object.defineProperty(exports, "parseDefinedTerm", {
  enumerable: true,
  get: function get() {
    return _DefinedTerm.parseDefinedTerm;
  }
});
Object.defineProperty(exports, "parseProject", {
  enumerable: true,
  get: function get() {
    return _Project.parseProject;
  }
});
Object.defineProperty(exports, "parseThing", {
  enumerable: true,
  get: function get() {
    return _Thing.parseThing;
  }
});
Object.defineProperty(exports, "parseVoteAction", {
  enumerable: true,
  get: function get() {
    return _VoteAction.parseVoteAction;
  }
});
Object.defineProperty(exports, "postActivity", {
  enumerable: true,
  get: function get() {
    return _postActivity["default"];
  }
});
Object.defineProperty(exports, "sendEvent", {
  enumerable: true,
  get: function get() {
    return _sendEvent["default"];
  }
});
Object.defineProperty(exports, "sendFiles", {
  enumerable: true,
  get: function get() {
    return _sendFiles["default"];
  }
});
Object.defineProperty(exports, "sendMessage", {
  enumerable: true,
  get: function get() {
    return _sendMessage["default"];
  }
});
Object.defineProperty(exports, "sendMessageBack", {
  enumerable: true,
  get: function get() {
    return _sendMessageBack["default"];
  }
});
Object.defineProperty(exports, "sendPostBack", {
  enumerable: true,
  get: function get() {
    return _sendPostBack["default"];
  }
});
Object.defineProperty(exports, "setDictateInterims", {
  enumerable: true,
  get: function get() {
    return _setDictateInterims["default"];
  }
});
Object.defineProperty(exports, "setDictateState", {
  enumerable: true,
  get: function get() {
    return _setDictateState["default"];
  }
});
Object.defineProperty(exports, "setLanguage", {
  enumerable: true,
  get: function get() {
    return _setLanguage["default"];
  }
});
Object.defineProperty(exports, "setNotification", {
  enumerable: true,
  get: function get() {
    return _setNotification["default"];
  }
});
Object.defineProperty(exports, "setSendBox", {
  enumerable: true,
  get: function get() {
    return _setSendBox["default"];
  }
});
Object.defineProperty(exports, "setSendBoxAttachments", {
  enumerable: true,
  get: function get() {
    return _setSendBoxAttachments["default"];
  }
});
Object.defineProperty(exports, "setSendTimeout", {
  enumerable: true,
  get: function get() {
    return _setSendTimeout["default"];
  }
});
Object.defineProperty(exports, "setSendTypingIndicator", {
  enumerable: true,
  get: function get() {
    return _setSendTypingIndicator["default"];
  }
});
Object.defineProperty(exports, "singleToArray", {
  enumerable: true,
  get: function get() {
    return _singleToArray["default"];
  }
});
Object.defineProperty(exports, "startDictate", {
  enumerable: true,
  get: function get() {
    return _startDictate["default"];
  }
});
Object.defineProperty(exports, "startSpeakingActivity", {
  enumerable: true,
  get: function get() {
    return _startSpeakingActivity["default"];
  }
});
Object.defineProperty(exports, "stopDictate", {
  enumerable: true,
  get: function get() {
    return _stopDictate["default"];
  }
});
Object.defineProperty(exports, "stopSpeakingActivity", {
  enumerable: true,
  get: function get() {
    return _stopSpeakingActivity["default"];
  }
});
Object.defineProperty(exports, "submitSendBox", {
  enumerable: true,
  get: function get() {
    return _submitSendBox["default"];
  }
});
exports.version = void 0;
Object.defineProperty(exports, "warnOnce", {
  enumerable: true,
  get: function get() {
    return _warnOnce["default"];
  }
});
var _clearSuggestedActions = _interopRequireDefault(require("./actions/clearSuggestedActions"));
var _connect = _interopRequireDefault(require("./actions/connect"));
var _disconnect = _interopRequireDefault(require("./actions/disconnect"));
var _dismissNotification = _interopRequireDefault(require("./actions/dismissNotification"));
var _emitTypingIndicator = _interopRequireDefault(require("./actions/emitTypingIndicator"));
var _markActivity = _interopRequireDefault(require("./actions/markActivity"));
var _postActivity = _interopRequireDefault(require("./actions/postActivity"));
var _sendEvent = _interopRequireDefault(require("./actions/sendEvent"));
var _sendFiles = _interopRequireDefault(require("./actions/sendFiles"));
var _sendMessage = _interopRequireDefault(require("./actions/sendMessage"));
var _sendMessageBack = _interopRequireDefault(require("./actions/sendMessageBack"));
var _sendPostBack = _interopRequireDefault(require("./actions/sendPostBack"));
var _setDictateInterims = _interopRequireDefault(require("./actions/setDictateInterims"));
var _setDictateState = _interopRequireDefault(require("./actions/setDictateState"));
var _setLanguage = _interopRequireDefault(require("./actions/setLanguage"));
var _setNotification = _interopRequireDefault(require("./actions/setNotification"));
var _setSendBox = _interopRequireDefault(require("./actions/setSendBox"));
var _setSendBoxAttachments = _interopRequireDefault(require("./actions/setSendBoxAttachments"));
var _setSendTimeout = _interopRequireDefault(require("./actions/setSendTimeout"));
var _setSendTypingIndicator = _interopRequireDefault(require("./actions/setSendTypingIndicator"));
var _startDictate = _interopRequireDefault(require("./actions/startDictate"));
var _startSpeakingActivity = _interopRequireDefault(require("./actions/startSpeakingActivity"));
var _stopDictate = _interopRequireDefault(require("./actions/stopDictate"));
var _stopSpeakingActivity = _interopRequireDefault(require("./actions/stopSpeakingActivity"));
var _submitSendBox = _interopRequireDefault(require("./actions/submitSendBox"));
var ActivityClientState = _interopRequireWildcard(require("./constants/ActivityClientState"));
var DictateState = _interopRequireWildcard(require("./constants/DictateState"));
var _createStore = _interopRequireWildcard(require("./createStore"));
var _Action = require("./types/external/OrgSchema/Action");
var _Claim = require("./types/external/OrgSchema/Claim");
var _CreativeWork = require("./types/external/OrgSchema/CreativeWork");
var _DefinedTerm = require("./types/external/OrgSchema/DefinedTerm");
var _Project = require("./types/external/OrgSchema/Project");
var _Thing = require("./types/external/OrgSchema/Thing");
var _VoteAction = require("./types/external/OrgSchema/VoteAction");
var _getOrgSchemaMessage = _interopRequireDefault(require("./utils/getOrgSchemaMessage"));
var _isForbiddenPropertyName = _interopRequireDefault(require("./utils/isForbiddenPropertyName"));
var _onErrorResumeNext = _interopRequireDefault(require("./utils/onErrorResumeNext"));
var _singleToArray = _interopRequireDefault(require("./utils/singleToArray"));
var _warnOnce = _interopRequireDefault(require("./utils/warnOnce"));
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
var Constants = {
  ActivityClientState: ActivityClientState,
  DictateState: DictateState
};
exports.Constants = Constants;
var buildTool = "babel";
var moduleFormat = "commonjs";
exports.moduleFormat = moduleFormat;
var version = "4.18.0";
exports.version = version;
var buildInfo = {
  buildTool: buildTool,
  moduleFormat: moduleFormat,
  version: version
};
exports.buildInfo = buildInfo;
