// botframework-webchat-api/lib/decorator/private/templateMiddleware.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = templateMiddleware;
var _botframeworkWebchatCore = require("botframework-webchat-core");
var _reactChainOfResponsibility = require("react-chain-of-responsibility");
var _valibot = require("valibot");
var EMPTY_ARRAY = Object.freeze([]);
function templateMiddleware(name) {
  var middlewareSchema = (0, _valibot.array)((0, _valibot.any)([(0, _valibot.custom)(function (input) {
    return typeof input === 'function';
  }, 'Middleware must be a function.')]));
  var isMiddleware = function isMiddleware(middleware) {
    return (0, _valibot.safeParse)(middlewareSchema, middleware).success;
  };
  var warnInvalid = (0, _botframeworkWebchatCore.warnOnce)("\"".concat(name, "\" prop is invalid"));
  var initMiddleware = function initMiddleware(middleware, init) {
    if (middleware) {
      if (isMiddleware(middleware)) {
        return Object.freeze(middleware === null || middleware === void 0 ? void 0 : middleware.map(function (middleware) {
          return middleware(init);
        }).filter(function (enhancer) {
          return !!enhancer;
        }).map(function (enhancer) {
          return function () {
            return enhancer;
          };
        }));
      }
      warnInvalid();
    }
    return EMPTY_ARRAY;
  };
  var _createChainOfRespons = (0, _reactChainOfResponsibility.createChainOfResponsibility)(),
    Provider = _createChainOfRespons.Provider,
    Proxy = _createChainOfRespons.Proxy;
  Provider.displayName = "".concat(name, "Provider");
  Proxy.displayName = "".concat(name, "Proxy");
  return {
    initMiddleware: initMiddleware,
    Provider: Provider,
    Proxy: Proxy,
    types: {
      init: undefined,
      middleware: undefined,
      props: undefined,
      request: undefined
    }
  };
}
