// botframework-webchat/lib/markdown/markdownItPlugins/ariaLabel.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.pre = exports.post = exports.default = void 0;
var _markdownItAttrsEs = _interopRequireDefault(require("markdown-it-attrs-es5"));
var MARKDOWN_ATTRS_LEFT_DELIMITER = '⟬';
// Make sure the delimiter is free from any RegExp characters, such as *, ?, etc.
// IE11 does not support "u" flag and Babel could not remove it. We intentionally omitting the "u" flag here.
// eslint-disable-next-line require-unicode-regexp
var MARKDOWN_ATTRS_LEFT_DELIMITER_PATTERN = new RegExp(MARKDOWN_ATTRS_LEFT_DELIMITER, 'g');
var MARKDOWN_ATTRS_RIGHT_DELIMITER = '⟭';
// Make sure the delimiter is free from any RegExp characters, such as *, ?, etc.
// IE11 does not support "u" flag and Babel could not remove it. We intentionally omitting the "u" flag here.
// eslint-disable-next-line require-unicode-regexp
var MARKDOWN_ATTRS_RIGHT_DELIMITER_PATTERN = new RegExp(MARKDOWN_ATTRS_RIGHT_DELIMITER, 'g');
var ariaLabel = function ariaLabel(markdown) {
  return markdown.use(_markdownItAttrsEs.default, {
    // `markdown-it-attrs` is added for accessibility and allow bot developers to specify `aria-label`.
    // We are allowlisting `aria-label` only as it is allowlisted in `sanitize-html`.
    // Other `aria-*` will be sanitized even we allowlisted here.
    allowedAttributes: ['aria-label'],
    leftDelimiter: MARKDOWN_ATTRS_LEFT_DELIMITER,
    rightDelimiter: MARKDOWN_ATTRS_RIGHT_DELIMITER
  });
};

// TODO: We should fold pre/post back into the plugin.
var pre = function pre(markdown) {
  return (
    // Related to #3165.
    // We only support attributes "aria-label" and should leave other attributes as-is.
    // However, `markdown-it-attrs` remove unrecognized attributes, such as {hello}.
    // Before passing to `markdown-it-attrs`, we will convert known attributes from {aria-label="..."} into ⟬aria-label="..."⟭ (using white tortoise shell brackets).
    // Then, we ask `markdown-it-attrs` to only process the new brackets, so it should only try to process things that we allowlisted.
    // Lastly, we revert tortoise shell brackets back to curly brackets, for unprocessed attributes.
    markdown
    // IE11 does not support "u" flag and Babel could not remove it. We intentionally omitting the "u" flag here.
    // eslint-disable-next-line require-unicode-regexp
    .replace(/\{\s*aria-label()\s*\}/gi, "".concat(MARKDOWN_ATTRS_LEFT_DELIMITER, "aria-label").concat(MARKDOWN_ATTRS_RIGHT_DELIMITER)).replace(
    // IE11 does not support "u" flag and Babel could not remove it. We intentionally omitting the "u" flag here.
    // eslint-disable-next-line require-unicode-regexp
    /\{\s*aria-label=("[^"]*"|[^\s}]*)\s*\}/gi, function (_, valueInsideQuotes) {
      return "".concat(MARKDOWN_ATTRS_LEFT_DELIMITER, "aria-label=").concat(valueInsideQuotes).concat(MARKDOWN_ATTRS_RIGHT_DELIMITER);
    })
  );
};
exports.pre = pre;
var post = function post(html) {
  return (
    // Restore attributes not processed by `markdown-it-attrs`.
    // TODO: [P2] #2511 After we fixed our polyfill story, we should use "String.prototype.replaceAll" instead of RegExp for replace all occurrences.
    html.replace(MARKDOWN_ATTRS_LEFT_DELIMITER_PATTERN, '{').replace(MARKDOWN_ATTRS_RIGHT_DELIMITER_PATTERN, '}')
  );
};
exports.post = post;
var _default = ariaLabel;
exports.default = _default;
