// botframework-webchat/lib/markdown/renderMarkdown.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
var _typeof = require("@babel/runtime/helpers/typeof");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = render;
var _botframeworkWebchatCore = require("botframework-webchat-core");
var _markdownIt = _interopRequireDefault(require("markdown-it"));
var _sanitizeHtml = _interopRequireDefault(require("sanitize-html"));
var _internal = require("botframework-webchat-component/internal");
var _ariaLabel = _interopRequireWildcard(require("./markdownItPlugins/ariaLabel"));
var _respectCRLF = require("./markdownItPlugins/respectCRLF");
var _betterLinkDocumentMod = _interopRequireDefault(require("./private/betterLinkDocumentMod"));
var _iterateLinkDefinitions = _interopRequireDefault(require("./private/iterateLinkDefinitions"));
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
var SANITIZE_HTML_OPTIONS = Object.freeze({
  allowedAttributes: {
    a: ['aria-label', 'class', 'href', 'name', 'rel', 'target'],
    button: ['aria-label', 'class', 'type', 'value'],
    img: ['alt', 'class', 'src', 'title'],
    span: ['aria-label']
  },
  allowedSchemes: ['data', 'http', 'https', 'ftp', 'mailto', 'sip', 'tel'],
  allowedTags: ['a', 'b', 'blockquote', 'br', 'button', 'caption', 'code', 'del', 'div', 'em', 'h1', 'h2', 'h3', 'h4', 'h5', 'h6', 'hr', 'i', 'img', 'ins', 'li', 'nl', 'ol', 'p', 'pre', 's', 'span', 'strike', 'strong', 'table', 'tbody', 'td', 'tfoot', 'th', 'thead', 'tr', 'ul'],
  // Bug of https://github.com/apostrophecms/sanitize-html/issues/633.
  // They should not remove `alt=""` even though it is empty.
  nonBooleanAttributes: []
});
function render(markdown, _ref) {
  var markdownRespectCRLF = _ref.markdownRespectCRLF,
    markdownRenderHTML = _ref.markdownRenderHTML;
  var _ref2 = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : Object.freeze({}),
    _ref2$externalLinkAlt = _ref2.externalLinkAlt,
    externalLinkAlt = _ref2$externalLinkAlt === void 0 ? '' : _ref2$externalLinkAlt;
  var linkDefinitions = Array.from((0, _iterateLinkDefinitions.default)(markdown));
  var MARKDOWN_IT_INIT = Object.freeze({
    breaks: false,
    html: markdownRenderHTML !== null && markdownRenderHTML !== void 0 ? markdownRenderHTML : true,
    linkify: true,
    typographer: true,
    xhtmlOut: true
  });
  if (markdownRespectCRLF) {
    markdown = (0, _respectCRLF.pre)(markdown);
  }
  markdown = (0, _ariaLabel.pre)(markdown);
  var decorate = function decorate(href, textContent) {
    var decoration = {
      rel: 'noopener noreferrer',
      target: '_blank',
      wrapZeroWidthSpace: true
    };
    var ariaLabelSegments = [textContent];
    var classes = new Set();
    var linkDefinition = linkDefinitions.find(function (_ref3) {
      var url = _ref3.url;
      return url === href;
    });
    var protocol = (0, _botframeworkWebchatCore.onErrorResumeNext)(function () {
      return new URL(href).protocol;
    });
    if (linkDefinition) {
      ariaLabelSegments.push(linkDefinition.title || (0, _botframeworkWebchatCore.onErrorResumeNext)(function () {
        return new URL(linkDefinition.url).host;
      }) || linkDefinition.url);

      // linkDefinition.identifier is uppercase, while linkDefinition.label is as-is.
      linkDefinition.label === textContent && classes.add('webchat__render-markdown__pure-identifier');
    }

    // For links that would be sanitized out, let's turn them into a button so we could handle them later.
    if (!SANITIZE_HTML_OPTIONS.allowedSchemes.map(function (scheme) {
      return "".concat(scheme, ":");
    }).includes(protocol)) {
      decoration.asButton = true;
      classes.add('webchat__render-markdown__citation');
    } else if (protocol === 'http:' || protocol === 'https:') {
      decoration.iconAlt = externalLinkAlt;
      decoration.iconClassName = 'webchat__render-markdown__external-link-icon';
      ariaLabelSegments.push(externalLinkAlt);
    }

    // The first segment is textContent. Putting textContent is aria-label is useless.
    if (ariaLabelSegments.length > 1) {
      // If "aria-label" is already applied, do not overwrite it.
      decoration.ariaLabel = function (value) {
        return value || ariaLabelSegments.join(' ');
      };
    }
    decoration.className = Array.from(classes).join(' ');

    // By default, Markdown-It will set "title" to the link title in link definition.

    // However, "title" may be narrated by screen reader:
    // - Edge
    //   - <a> will narrate "aria-label" but not "title"
    //   - <button> will narrate both "aria-label" and "title"
    // - NVDA
    //   - <a> will narrate both "aria-label" and "title"
    //   - <button> will narrate both "aria-label" and "title"

    // Title makes it very difficult to control narrations by the screen reader. Thus, we are disabling it in favor of "aria-label".
    // This will not affect our accessibility compliance but UX. We could use a non-native tooltip or other forms of visual hint.

    decoration.title = false;
    return decoration;
  };
  var htmlAfterMarkdown = (0, _ariaLabel.post)(new _markdownIt.default(MARKDOWN_IT_INIT).use(_ariaLabel.default).render(markdown));

  // TODO: [P1] In some future, we should apply "better link" and "sanitization" outside of the Markdown engine.
  //       Particularly, apply them at `useRenderMarkdownAsHTML` instead of inside the default `renderMarkdown`.
  //       If web devs want to bring their own Markdown engine, they don't need to rebuild "better link" and sanitization themselves.

  var documentAfterMarkdown = (0, _internal.parseDocumentFromString)(htmlAfterMarkdown);
  (0, _betterLinkDocumentMod.default)(documentAfterMarkdown, decorate);
  var htmlAfterBetterLink = (0, _internal.serializeDocumentIntoString)(documentAfterMarkdown);
  var htmlAfterSanitization = (0, _sanitizeHtml.default)(htmlAfterBetterLink, SANITIZE_HTML_OPTIONS);
  return htmlAfterSanitization;
}
