// botframework-webchat-component/lib/Styles/StyleSet/CarouselFilmStrip.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = CarouselFilmStrip;
var _mirrorStyle = _interopRequireDefault(require("../mirrorStyle"));
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }
function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; } /* eslint no-magic-numbers: ["error", { "ignore": [2] }] */
function CarouselFilmStrip(_ref) {
  var avatarSize = _ref.avatarSize,
    bubbleMaxWidth = _ref.bubbleMaxWidth,
    paddingRegular = _ref.paddingRegular,
    transitionDuration = _ref.transitionDuration;
  return {
    '&.webchat__carousel-filmstrip': _objectSpread({
      // Browser quirks: Firefox has no way to hide scrollbar and while keeping it in function
      // https://developer.mozilla.org/en-US/docs/Web/CSS/overflow
      '@supports (-moz-appearance: none)': {
        marginBottom: -17
      },
      '& .webchat__carousel-filmstrip__message': {
        maxWidth: bubbleMaxWidth,
        transitionDuration: transitionDuration,
        transitionProperty: 'max-width'
      },
      '&.webchat__carousel-filmstrip--hide-nub, &.webchat__carousel-filmstrip--show-nub, &.webchat__carousel-filmstrip--hide-avatar, &.webchat__carousel-filmstrip--show-avatar': {
        '& .webchat__carousel-filmstrip__message': {
          maxWidth: bubbleMaxWidth + paddingRegular
        }
      },
      '& .webchat__carousel-filmstrip__alignment-pad': {
        transitionDuration: transitionDuration,
        transitionProperty: 'width',
        width: paddingRegular
      },
      '&.webchat__carousel-filmstrip--extra-trailing .webchat__carousel-filmstrip__alignment-pad': {
        width: paddingRegular * 2
      },
      '&:not(.webchat__carousel-filmstrip--no-message) .webchat__carousel-filmstrip__attachments': {
        marginTop: paddingRegular
      },
      '& .webchat__carousel-filmstrip__avatar-gutter': {
        alignItems: 'flex-end',
        transitionDuration: transitionDuration,
        transitionProperty: 'width'
      },
      '& .webchat__carousel-filmstrip__nub-pad': {
        transitionDuration: transitionDuration,
        transitionProperty: 'width',
        width: 0
      },
      '&.webchat__carousel-filmstrip--hide-avatar, &.webchat__carousel-filmstrip--show-avatar': {
        '& .webchat__carousel-filmstrip__avatar-gutter': {
          width: avatarSize
        }
      },
      '&.webchat__carousel-filmstrip--hide-avatar, &.webchat__carousel-filmstrip--show-avatar, &.webchat__carousel-filmstrip--hide-nub, &.webchat__carousel-filmstrip--show-nub': {
        '& .webchat__carousel-filmstrip__nub-pad': {
          width: paddingRegular
        }
      },
      '&:not(.webchat__carousel-filmstrip--top-callout) .webchat__carousel-filmstrip__avatar-gutter': {
        justifyContent: 'flex-end'
      }
    }, (0, _mirrorStyle.default)('&.webchat__carousel-filmstrip--rtl', {
      '& .webchat__carousel-filmstrip__avatar-gutter': {
        marginLeft: paddingRegular
      },
      '& .webchat__carousel-filmstrip__attachments': {
        marginLeft: -paddingRegular
      },
      '&.webchat__carousel-filmstrip--hide-avatar, &.webchat__carousel-filmstrip--show-avatar': {
        '& .webchat__carousel-filmstrip__attachments': {
          marginLeft: -(avatarSize + paddingRegular * 2)
        }
      },
      '&.webchat__carousel-filmstrip--hide-nub, &.webchat__carousel-filmstrip--show-nub': {
        '&:not(.webchat__carousel-filmstrip--hide-avatar.webchat__carousel-filmstrip--show-avatar)': {
          '& .webchat__carousel-filmstrip__attachments': {
            marginLeft: -paddingRegular * 2
          }
        }
      }
    }))
  };
}
