// botframework-webchat-component/lib/Styles/StyleSet/LinkDefinitions.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = createLinkDefinitionsStyleSet;
var _Constants = require("./Constants");
var _CSSTokens = _interopRequireDefault(require("../CSSTokens"));
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }
// TODO: Temporarily disable dark theme until chat history support dark theme.
var DARK_THEME_SELECTOR = '@media (forced-colors: none) and not (forced-colors: none)'; // Always return false
var LIGHT_THEME_SELECTOR = '@media (forced-colors: none)';
function createLinkDefinitionsStyleSet() {
  var _webchat__linkDefin, _webchat__linkDefin2, _webchat__linkDefin3;
  return {
    '&.webchat__link-definitions': {
      // '&[open] .webchat__link-definitions__header::after': {
      //   transform: 'rotate(0deg)'
      // },

      '.webchat__link-definitions__header': (_webchat__linkDefin = {
        alignItems: 'center',
        display: 'flex',
        fontFamily: _CSSTokens.default.FontPrimary,
        fontSize: _CSSTokens.default.FontSizeSmall,
        gap: 4,
        listStyle: 'none'
      }, _defineProperty(_webchat__linkDefin, LIGHT_THEME_SELECTOR, {
        color: '#616161' // TODO: Should we use subtle color instead?
      }), _defineProperty(_webchat__linkDefin, DARK_THEME_SELECTOR, {
        // TODO: Add dark theme color.
        color: '#616161'
      }), _webchat__linkDefin),
      '.webchat__link-definitions__header-text': {
        flexShrink: 0
      },
      '.webchat__link-definitions__header::-webkit-details-marker': {
        display: 'none'
      },
      '.webchat__link-definitions__header-chevron': {
        flexShrink: 0
      },
      '&:not([open]) .webchat__link-definitions__header-chevron': {
        transform: 'rotate(-180deg)'
      },
      '.webchat__link-definitions__header-filler': {
        flexGrow: 1
      },
      '.webchat__link-definitions__header-accessory': {
        overflow: 'hidden'
      },
      '.webchat__link-definitions__message-sensitivity-label': {
        display: 'flex',
        gap: 4
      },
      '.webchat__link-definitions__message-sensitivity-label-icon': {
        color: 'CanvasText',
        flexShrink: 0
      },
      '.webchat__link-definitions__message-sensitivity-label-text': {
        overflow: 'hidden',
        textOverflow: 'ellipsis',
        whiteSpace: 'nowrap'
      },
      '.webchat__link-definitions__list': {
        display: 'flex',
        flexDirection: 'column',
        gap: 4,
        margin: '4px 0 0',
        padding: 0
      },
      '.webchat__link-definitions__list-item': {
        display: 'flex',
        // This prevents the <button> from overflowing. Unsure why "overflow: hidden" doesn't work.
        flexDirection: 'column'
      },
      '.webchat__link-definitions__badge': (_webchat__linkDefin2 = {
        alignItems: 'center',
        alignSelf: 'flex-start',
        borderRadius: '4px',
        borderStyle: 'solid',
        borderWidth: 1,
        display: 'flex',
        flexShrink: 0,
        fontSize: '75%',
        justifyContent: 'center',
        margin: 4,
        minWidth: '1em',
        overflow: 'hidden',
        padding: 2,
        whiteSpace: 'nowrap'
      }, _defineProperty(_webchat__linkDefin2, LIGHT_THEME_SELECTOR, {
        backgroundColor: 'white',
        borderColor: '#e0e0e0',
        color: 'black'
      }), _defineProperty(_webchat__linkDefin2, DARK_THEME_SELECTOR, {
        backgroundColor: 'black',
        // TODO: Add dark theme color.
        borderColor: '#e0e0e0',
        color: 'white'
      }), _defineProperty(_webchat__linkDefin2, _Constants.FORCED_COLORS_SELECTOR, {
        borderColor: 'buttonborder'
      }), _webchat__linkDefin2),
      '.webchat__link-definitions__list-item-box': (_webchat__linkDefin3 = {
        alignItems: 'center',
        borderRadius: 4,
        borderStyle: 'solid',
        borderWidth: 1
      }, _defineProperty(_webchat__linkDefin3, LIGHT_THEME_SELECTOR, {
        backgroundColor: 'white',
        borderColor: '#d1d1d1'
      }), _defineProperty(_webchat__linkDefin3, DARK_THEME_SELECTOR, {
        backgroundColor: 'black',
        // TODO: Add dark theme color.
        borderColor: '#d1d1d1'
      }), _defineProperty(_webchat__linkDefin3, _Constants.FORCED_COLORS_SELECTOR, {
        backgroundColor: 'canvas',
        borderColor: 'buttonborder'
      }), _webchat__linkDefin3),
      '.webchat__link-definitions__list-item-box--as-link': {
        display: 'block',
        outlineOffset: 0,
        // This will make sure focus indicator is same as <button>.
        textDecoration: 'none'
      },
      '.webchat__link-definitions__list-item-box--as-button': {
        appearance: 'none',
        background: 'transparent',
        cursor: 'pointer',
        fontFamily: 'inherit',
        fontSize: 'inherit',
        overflow: 'hidden',
        padding: 0,
        textAlign: 'initial' // By default, texts inside button are centered.
      },
      '.webchat__link-definitions__list-item-body': {
        alignItems: 'center',
        display: 'flex',
        fontFamily: "Calibri, 'Helvetica Neue', Arial, 'sans-serif'",
        gap: 4,
        padding: 4
      },
      '.webchat__link-definitions__list-item-body-main': {
        display: 'flex',
        flexDirection: 'column',
        gap: 2,
        overflow: 'hidden'
      },
      '.webchat__link-definitions__list-item-main-text': {
        alignItems: 'baseline',
        display: 'flex',
        gap: 4
      },
      '.webchat__link-definitions__list-item-badge, .webchat__link-definitions__list-item-text': {
        overflow: 'hidden',
        textOverflow: 'ellipsis',
        whiteSpace: 'nowrap'
      },
      '.webchat__link-definitions__list-item-text': _defineProperty({
        textDecoration: 'underline'
      }, _Constants.NOT_FORCED_COLORS_SELECTOR, {
        color: _CSSTokens.default.ColorAccent
      }),
      '.webchat__link-definitions__list-item-badge': _defineProperty({
        fontSize: _CSSTokens.default.FontSizeSmall
      }, _Constants.NOT_FORCED_COLORS_SELECTOR, {
        color: _CSSTokens.default.ColorSubtle
      }),
      '.webchat__link-definitions__open-in-new-window-icon': _defineProperty({
        flexShrink: 0,
        // When text is too long, make sure the chevron is not squeezed.
        paddingRight: 4
      }, _Constants.NOT_FORCED_COLORS_SELECTOR, {
        color: _CSSTokens.default.ColorAccent
      })
    }
  };
}
