// botframework-webchat-component/lib/Styles/StyleSet/ModalDialog.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = createModalDialogStyleSet;
var _Constants = require("./Constants");
var _CSSTokens = _interopRequireDefault(require("../CSSTokens"));
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }
// TODO: Temporarily disable dark theme until we defined the link color and stuff for Markdown.
var DARK_THEME_SELECTOR = '@media (forced-colors: none) and not (forced-colors: none)'; // Always return false
var LIGHT_THEME_SELECTOR = '@media (forced-colors: none)';
function createModalDialogStyleSet() {
  var _webchat__modalDi, _webchat__modalDi3, _webchat__modalDi4, _webchat__modalDia;
  return {
    '&.webchat__modal-dialog': (_webchat__modalDia = {
      fontFamily: _CSSTokens.default.FontPrimary,
      maxHeight: 'calc(100% - 32px)',
      maxWidth: 'calc(100% - 32px)',
      width: '100%'
    }, _defineProperty(_webchat__modalDia, _Constants.NOT_FORCED_COLORS_SELECTOR, {
      backgroundColor: 'transparent',
      border: 0
    }), _defineProperty(_webchat__modalDia, '& .webchat__modal-dialog__box', (_webchat__modalDi = {
      borderRadius: 2,
      overflow: 'hidden',
      margin: 'auto',
      '@media screen and (max-width: 639px)': {
        maxWidth: 'unset'
      },
      '@media screen and (min-width: 640px)': {
        maxWidth: '60%',
        minWidth: 'calc(640px - 32px)',
        width: '60%'
      }
    }, _defineProperty(_webchat__modalDi, LIGHT_THEME_SELECTOR, {
      // From Power BI:
      // boxShadow: '0 6.4px 14.4px rgb(0 0 0 / 13%), 0 1.2px 3.6px rgb(0 0 0 / 11%)',
      // From Fluent (depth-16 for teaching callouts):
      // boxShadow: '0px 8px 16px rgba(0, 0, 0, 0.14), 0px 0px 2px rgba(0, 0, 0, 0.12)',
      backgroundColor: 'White',
      boxShadow: '0px 8px 16px rgba(0, 0, 0, 0.14), 0px 0px 2px rgba(0, 0, 0, 0.12)'
    }), _defineProperty(_webchat__modalDi, DARK_THEME_SELECTOR, {
      backgroundColor: 'Black',
      boxShadow: '0px 8px 16px rgba(0, 0, 0, 0.28), 0px 0px 2px rgba(0, 0, 0, 0.24)',
      color: 'White'
    }), _defineProperty(_webchat__modalDi, _Constants.FORCED_COLORS_SELECTOR, {
      // In high-contrast mode, we use "outline" instead of "box-shadow".
      outlineColor: 'ButtonBorder',
      outlineStyle: 'solid',
      outlineWidth: 4
    }), _webchat__modalDi)), _defineProperty(_webchat__modalDia, '& .webchat__modal-dialog__close-button-layout', {
      float: 'right',
      padding: _CSSTokens.default.PaddingRegular
    }), _defineProperty(_webchat__modalDia, '& .webchat__modal-dialog__close-button', _defineProperty({
      height: 30,
      width: 30
    }, _Constants.NOT_FORCED_COLORS_SELECTOR, {
      appearance: 'none',
      backgroundColor: 'transparent',
      borderColor: 'transparent',
      borderRadius: 4,
      borderStyle: 'solid',
      borderWidth: 2,
      color: '#999',
      outline: 0
    })), _defineProperty(_webchat__modalDia, '& .webchat__modal-dialog__close-button:focus', (_webchat__modalDi3 = {}, _defineProperty(_webchat__modalDi3, LIGHT_THEME_SELECTOR, {
      borderColor: 'black',
      '&:active': {
        backgroundColor: '#EDEBE9' // neutralLight (gray30)
      },
      '&:not(:active):hover': {
        backgroundColor: '#F3F2F1' // neutralLighter (gray20)
      }
    }), _defineProperty(_webchat__modalDi3, DARK_THEME_SELECTOR, {
      borderColor: 'white',
      '&:active': {
        backgroundColor: '#292827' // neutralLight (gray30)
      },
      '&:not(:active):hover': {
        backgroundColor: '#252423' // neutralLight (gray30)
      }
    }), _webchat__modalDi3)), _defineProperty(_webchat__modalDia, '& .webchat__modal-dialog__close-button-image', (_webchat__modalDi4 = {
      height: 10,
      width: 10
    }, _defineProperty(_webchat__modalDi4, LIGHT_THEME_SELECTOR, {
      fill: '#323130' // neutralPrimary (gray160)
    }), _defineProperty(_webchat__modalDi4, DARK_THEME_SELECTOR, {
      fill: '#F3F2F1' // neutralPrimary (gray160)
    }), _defineProperty(_webchat__modalDi4, _Constants.FORCED_COLORS_SELECTOR, {
      fill: 'currentcolor'
    }), _webchat__modalDi4)), _defineProperty(_webchat__modalDia, '& .webchat__modal-dialog__body', {
      margin: "calc(".concat(_CSSTokens.default.PaddingRegular, " * 2)")
    }), _webchat__modalDia)
  };
}
