// botframework-webchat-component/lib/Styles/StyleSet/RenderMarkdown.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = createMarkdownStyle;
var _CSSTokens = _interopRequireDefault(require("../CSSTokens"));
var _Constants = require("./Constants");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }
// This style is for accompanying result of `renderMarkdown()`.
// Mostly, it should only styles elements that are generated/modified during `renderMarkdown()`.
// For example, "open in new window" icon, which is done by `betterLink`.
function createMarkdownStyle() {
  var _webchat__renderM;
  return {
    '&.webchat__render-markdown': {
      display: 'contents',
      // Copied from Adaptive Cards inline styling.
      '&.webchat__render-markdown--adaptive-cards > *:first-child': {
        marginTop: 0,
        width: '100%',
        overflow: 'hidden',
        textOverflow: 'ellipsis'
      },
      // Copied from Adaptive Cards inline styling.
      '&.webchat__render-markdown--adaptive-cards > *:last-child': {
        marginBottom: 0
      },
      '&.webchat__render-markdown--message-activity > *:first-child': {
        marginTop: 0
      },
      '&.webchat__render-markdown--message-activity > *:last-child': {
        marginBottom: 0
      },
      '& .webchat__render-markdown__external-link-icon': {
        backgroundImage: _CSSTokens.default.IconURLExternalLink,
        height: '.75em',
        marginLeft: '.25em'
      },
      '& .webchat__render-markdown__citation': (_webchat__renderM = {
        background: 'transparent',
        border: 0,
        cursor: 'pointer',
        fontFamily: 'unset',
        fontSize: 'unset',
        padding: 0,
        textDecoration: 'underline',
        whiteSpace: 'nowrap'
      }, _defineProperty(_webchat__renderM, _Constants.FORCED_COLORS_SELECTOR, {
        color: 'LinkText'
      }), _defineProperty(_webchat__renderM, _Constants.NOT_FORCED_COLORS_SELECTOR, {
        color: _CSSTokens.default.ColorAccent
      }), _webchat__renderM),
      '& .webchat__render-markdown__pure-identifier': {
        whiteSpace: 'nowrap'
      },
      '& .webchat__render-markdown__pure-identifier::after': {
        content: "']'"
      },
      '& .webchat__render-markdown__pure-identifier::before': {
        content: "'['"
      }
    }
  };
}
