// botframework-webchat-component/lib/Styles/StyleSet/TextContent.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = createTextContentStyle;
var _CSSTokens = _interopRequireDefault(require("../CSSTokens"));
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function createTextContentStyle() {
  return {
    '&.webchat__text-content': {
      fontFamily: _CSSTokens.default.FontPrimary,
      margin: 0,
      minHeight: "calc(".concat(_CSSTokens.default.MinHeightBubble, " - ").concat(_CSSTokens.default.PaddingRegular, " * 2)"),
      padding: _CSSTokens.default.PaddingRegular,
      '&.webchat__text-content--is-markdown': {
        display: 'flex',
        flexDirection: 'column',
        gap: _CSSTokens.default.PaddingRegular
      },
      '& .webchat__text-content__markdown img:not(.webchat__render-markdown__external-link-icon)': {
        maxWidth: _CSSTokens.default.MaxWidthBubble,
        width: '100%'
      },
      '& .webchat__text-content__markdown pre': {
        overflow: 'hidden'
      },
      '& .webchat__text-content__open-in-new-window-icon': {
        height: '.75em'
      }
    }
  };
}
