// botframework-webchat-core/lib/sagas/sendMessageToPostActivitySaga.js
"use strict";

var _regeneratorRuntime2 = require("@babel/runtime/regenerator");
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = sendMessageToPostActivitySaga;
var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _effects = require("redux-saga/effects");
var _postActivity = _interopRequireDefault(require("../actions/postActivity"));
var _sendMessage = require("../actions/sendMessage");
var _whileConnected = _interopRequireDefault(require("./effects/whileConnected"));
var _marked = /*#__PURE__*/_regeneratorRuntime2.mark(postActivityWithMessage),
  _marked2 = /*#__PURE__*/_regeneratorRuntime2.mark(sendMessageToPostActivity),
  _marked3 = /*#__PURE__*/_regeneratorRuntime2.mark(sendMessageToPostActivitySaga);
function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }
function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { (0, _defineProperty2["default"])(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
function postActivityWithMessage(_ref) {
  var _ref$payload, _ref$payload$attachme, attachments, channelData, method, text;
  return _regenerator["default"].wrap(function postActivityWithMessage$(_context) {
    while (1) {
      switch (_context.prev = _context.next) {
        case 0:
          _ref$payload = _ref.payload, _ref$payload$attachme = _ref$payload.attachments, attachments = _ref$payload$attachme === void 0 ? [] : _ref$payload$attachme, channelData = _ref$payload.channelData, method = _ref$payload.method, text = _ref$payload.text;
          _context.next = 3;
          return (0, _effects.put)((0, _postActivity["default"])({
            attachments: attachments.map(function (_ref2) {
              var blob = _ref2.blob,
                thumbnailURL = _ref2.thumbnailURL;
              return {
                contentType: blob instanceof File && blob.type || 'application/octet-stream',
                contentUrl: URL.createObjectURL(blob),
                name: blob instanceof File ? blob.name : undefined,
                thumbnailUrl: thumbnailURL === null || thumbnailURL === void 0 ? void 0 : thumbnailURL.toString()
              };
            }),
            channelData: _objectSpread(_objectSpread({}, channelData), {}, {
              attachmentSizes: attachments.map(function (_ref3) {
                var size = _ref3.blob.size;
                return size;
              })
            }),
            text: text || undefined,
            textFormat: 'plain',
            type: 'message'
          },
          // TODO: Fix WebChatActivity. Currently, it only works with incoming activity, not outgoing.
          method));
        case 3:
        case "end":
          return _context.stop();
      }
    }
  }, _marked);
}
function sendMessageToPostActivity() {
  return _regenerator["default"].wrap(function sendMessageToPostActivity$(_context2) {
    while (1) {
      switch (_context2.prev = _context2.next) {
        case 0:
          _context2.next = 2;
          return (0, _effects.takeEvery)(function (_ref4) {
            var _payload$attachments;
            var payload = _ref4.payload,
              type = _ref4.type;
            return type === _sendMessage.SEND_MESSAGE && (payload.text || ((_payload$attachments = payload.attachments) === null || _payload$attachments === void 0 ? void 0 : _payload$attachments.length));
          }, postActivityWithMessage);
        case 2:
        case "end":
          return _context2.stop();
      }
    }
  }, _marked2);
}
function sendMessageToPostActivitySaga() {
  return _regenerator["default"].wrap(function sendMessageToPostActivitySaga$(_context3) {
    while (1) {
      switch (_context3.prev = _context3.next) {
        case 0:
          _context3.next = 2;
          return (0, _whileConnected["default"])(sendMessageToPostActivity);
        case 2:
        case "end":
          return _context3.stop();
      }
    }
  }, _marked3);
}
