// botframework-webchat-core/lib/sagas/submitSendBoxSaga.js
"use strict";

var _regeneratorRuntime2 = require("@babel/runtime/regenerator");
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = submitSendBoxSaga;
var _regenerator = _interopRequireDefault(require("@babel/runtime/regenerator"));
var _effects = require("redux-saga/effects");
var _sendMessage = _interopRequireDefault(require("../actions/sendMessage"));
var _setSendBox = _interopRequireDefault(require("../actions/setSendBox"));
var _setSendBoxAttachments = _interopRequireDefault(require("../actions/setSendBoxAttachments"));
var _submitSendBox = require("../actions/submitSendBox");
var _sendBoxAttachments = _interopRequireDefault(require("../selectors/sendBoxAttachments"));
var _sendBoxValue = _interopRequireDefault(require("../selectors/sendBoxValue"));
var _whileConnected = _interopRequireDefault(require("./effects/whileConnected"));
var _marked = /*#__PURE__*/_regeneratorRuntime2.mark(submitSendBox),
  _marked2 = /*#__PURE__*/_regeneratorRuntime2.mark(submitSendBoxSaga);
function submitSendBox() {
  return _regenerator["default"].wrap(function submitSendBox$(_context2) {
    while (1) {
      switch (_context2.prev = _context2.next) {
        case 0:
          _context2.next = 2;
          return (0, _effects.takeEvery)(_submitSendBox.SUBMIT_SEND_BOX, /*#__PURE__*/_regenerator["default"].mark(function _callee(_ref) {
            var _ref$payload, channelData, method, attachments, sendBoxValue;
            return _regenerator["default"].wrap(function _callee$(_context) {
              while (1) {
                switch (_context.prev = _context.next) {
                  case 0:
                    _ref$payload = _ref.payload, channelData = _ref$payload.channelData, method = _ref$payload.method;
                    _context.next = 3;
                    return (0, _effects.select)(_sendBoxAttachments["default"]);
                  case 3:
                    attachments = _context.sent;
                    _context.next = 6;
                    return (0, _effects.select)(_sendBoxValue["default"]);
                  case 6:
                    sendBoxValue = _context.sent;
                    if (!(attachments.length || sendBoxValue)) {
                      _context.next = 14;
                      break;
                    }
                    _context.next = 10;
                    return (0, _effects.put)((0, _sendMessage["default"])(sendBoxValue.trim(), method, {
                      attachments: attachments,
                      channelData: channelData
                    }));
                  case 10:
                    _context.next = 12;
                    return (0, _effects.put)((0, _setSendBox["default"])(''));
                  case 12:
                    _context.next = 14;
                    return (0, _effects.put)((0, _setSendBoxAttachments["default"])(Object.freeze([])));
                  case 14:
                  case "end":
                    return _context.stop();
                }
              }
            }, _callee);
          }));
        case 2:
        case "end":
          return _context2.stop();
      }
    }
  }, _marked);
}
function submitSendBoxSaga() {
  return _regenerator["default"].wrap(function submitSendBoxSaga$(_context3) {
    while (1) {
      switch (_context3.prev = _context3.next) {
        case 0:
          _context3.next = 2;
          return (0, _whileConnected["default"])(submitSendBox);
        case 2:
        case "end":
          return _context3.stop();
      }
    }
  }, _marked2);
}
