// botframework-webchat-component/lib/Utils/betterLinks.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _markdownIt = _interopRequireDefault(require("markdown-it"));
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function _toConsumableArray(arr) { return _arrayWithoutHoles(arr) || _iterableToArray(arr) || _unsupportedIterableToArray(arr) || _nonIterableSpread(); }
function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }
function _iterableToArray(iter) { if (typeof Symbol !== "undefined" && iter[Symbol.iterator] != null || iter["@@iterator"] != null) return Array.from(iter); }
function _arrayWithoutHoles(arr) { if (Array.isArray(arr)) return _arrayLikeToArray(arr); }
function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }
function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }
function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }
function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }
function _iterableToArrayLimit(arr, i) { var _i = arr == null ? null : typeof Symbol !== "undefined" && arr[Symbol.iterator] || arr["@@iterator"]; if (_i == null) return; var _arr = []; var _n = true; var _d = false; var _s, _e; try { for (_i = _i.call(arr); !(_n = (_s = _i.next()).done); _n = true) { _arr.push(_s.value); if (i && _arr.length === i) break; } } catch (err) { _d = true; _e = err; } finally { try { if (!_n && _i["return"] != null) _i["return"](); } finally { if (_d) throw _e; } } return _arr; }
function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; } /* eslint-disable security/detect-object-injection */
function iterator(md, ruleName, tokenType, iterator) {
  function scan(state) {
    var _state$env = state.env,
      env = _state$env === void 0 ? {} : _state$env;
    for (var blkIdx = state.tokens.length - 1; blkIdx >= 0; blkIdx--) {
      if (state.tokens[blkIdx].type !== 'inline') {
        continue;
      }
      var inlineTokens = state.tokens[blkIdx].children;
      for (var i = inlineTokens.length - 1; i >= 0; i--) {
        if (inlineTokens[i].type !== tokenType) {
          continue;
        }
        iterator(inlineTokens, i, env);
      }
    }
  }
  md.core.ruler.push(ruleName, scan);
}

// Put a transparent pixel instead of the "open in new window" icon, so developers can easily modify the icon in CSS.
var TRANSPARENT_GIF = 'data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7';
// This is used for parsing Markdown for external links.
var internalMarkdownIt = new _markdownIt.default();
var ZERO_WIDTH_SPACE_TOKEN = {
  content: "\u200B",
  type: 'text'
};
function setTokenAttribute(attrs, name, value) {
  var index = attrs.findIndex(function (entry) {
    return entry[0] === name;
  });
  if (value === false) {
    ~index && attrs.splice(index, 1);
  } else if (typeof value === 'string') {
    if (~index) {
      attrs[+index][1] = value;
    } else {
      attrs.push([name, value]);
    }
  } else if (typeof value === 'function') {
    if (~index) {
      attrs[+index][1] = value(attrs[+index][1]);
    } else {
      attrs.push([name, value()]);
    }
  }
}
var betterLinks = function betterLinks(markdown) {
  return markdown.use(iterator, 'url_new_win', 'link_open', function (tokens, index, env) {
    var decorateLink = env.decorateLink,
      linkOptions = env.linkOptions;
    var indexOfLinkCloseToken = tokens.indexOf(tokens.slice(index + 1).find(function (_ref) {
      var type = _ref.type;
      return type === 'link_close';
    }));
    // eslint-disable-next-line no-magic-numbers
    var updatedTokens = tokens.splice(index, ~indexOfLinkCloseToken ? indexOfLinkCloseToken - index + 1 : 2);
    try {
      var _updatedTokens = _slicedToArray(updatedTokens, 1),
        linkOpenToken = _updatedTokens[0];
      var linkCloseToken = updatedTokens[updatedTokens.length - 1];
      var _linkOpenToken$attrs$ = linkOpenToken.attrs.find(function (_ref2) {
          var _ref3 = _slicedToArray(_ref2, 1),
            name = _ref3[0];
          return name === 'href';
        }),
        _linkOpenToken$attrs$2 = _slicedToArray(_linkOpenToken$attrs$, 2),
        _href = _linkOpenToken$attrs$2[1];
      var nodesInLink = updatedTokens.slice(1, updatedTokens.length - 1);
      var _textContent = nodesInLink.filter(function (_ref4) {
        var type = _ref4.type;
        return type === 'text';
      }).map(function (_ref5) {
        var content = _ref5.content;
        return content;
      }).join(' ');
      var decoration = decorateLink === null || decorateLink === void 0 ? void 0 : decorateLink(_href, _textContent, linkOptions);
      if (!decoration) {
        return;
      }
      var ariaLabel = decoration.ariaLabel,
        asButton = decoration.asButton,
        className = decoration.className,
        iconAlt = decoration.iconAlt,
        iconClassName = decoration.iconClassName,
        rel = decoration.rel,
        target = decoration.target,
        title = decoration.title,
        wrapZeroWidthSpace = decoration.wrapZeroWidthSpace;
      setTokenAttribute(linkOpenToken.attrs, 'aria-label', ariaLabel);
      setTokenAttribute(linkOpenToken.attrs, 'class', className);
      setTokenAttribute(linkOpenToken.attrs, 'title', title);
      if (iconClassName) {
        var iconTokens = internalMarkdownIt.parseInline("![](".concat(TRANSPARENT_GIF, ")"))[0].children;
        setTokenAttribute(iconTokens[0].attrs, 'class', iconClassName);
        setTokenAttribute(iconTokens[0].attrs, 'title', iconAlt);

        // Add an icon before </a>.
        // eslint-disable-next-line no-magic-numbers
        updatedTokens.splice.apply(updatedTokens, [-1, 0].concat(_toConsumableArray(iconTokens)));
      }
      if (asButton) {
        setTokenAttribute(linkOpenToken.attrs, 'href', false);
        linkOpenToken.tag = 'button';
        setTokenAttribute(linkOpenToken.attrs, 'type', 'button');
        setTokenAttribute(linkOpenToken.attrs, 'value', _href);
        linkCloseToken.tag = 'button';
      } else {
        setTokenAttribute(linkOpenToken.attrs, 'rel', rel);
        setTokenAttribute(linkOpenToken.attrs, 'target', target);
      }
      if (wrapZeroWidthSpace) {
        updatedTokens.splice(0, 0, ZERO_WIDTH_SPACE_TOKEN);
        updatedTokens.splice(Infinity, 0, ZERO_WIDTH_SPACE_TOKEN);
      }
    } finally {
      tokens.splice.apply(tokens, [index, 0].concat(_toConsumableArray(updatedTokens)));
    }
  });
};
var _default = betterLinks;
exports.default = _default;
