// botframework-webchat-component/lib/Utils/LocalizedString.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; return _typeof = "function" == typeof Symbol && "symbol" == typeof Symbol.iterator ? function (obj) { return typeof obj; } : function (obj) { return obj && "function" == typeof Symbol && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }, _typeof(obj); }
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _botframeworkWebchatApi = require("botframework-webchat-api");
var _botframeworkWebchatCore = require("botframework-webchat-core");
var _markdownIt = _interopRequireDefault(require("markdown-it"));
var _react = _interopRequireWildcard(require("react"));
var _betterLinks = _interopRequireDefault(require("./betterLinks"));
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function _toConsumableArray(arr) { return _arrayWithoutHoles(arr) || _iterableToArray(arr) || _unsupportedIterableToArray(arr) || _nonIterableSpread(); }
function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }
function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }
function _iterableToArray(iter) { if (typeof Symbol !== "undefined" && iter[Symbol.iterator] != null || iter["@@iterator"] != null) return Array.from(iter); }
function _arrayWithoutHoles(arr) { if (Array.isArray(arr)) return _arrayLikeToArray(arr); }
function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }
function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }
function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; } /* eslint react/no-danger: "off" */
var allowedSchemes = ['data', 'http', 'https', 'ftp', 'mailto', 'sip', 'tel'];
var linkDefinitions = [];
var externalLinkAlt = '';
var defaultDecorateLink = function defaultDecorateLink(href, textContent, linkOptions) {
  var _decoration$title;
  var decoration = _objectSpread({
    rel: 'noopener noreferrer',
    target: '_blank',
    wrapZeroWidthSpace: true
  }, linkOptions);
  var ariaLabelSegments = [textContent];
  var classes = new Set();
  var linkDefinition = linkDefinitions.find(function (_ref) {
    var url = _ref.url;
    return url === href;
  });
  var protocol = (0, _botframeworkWebchatCore.onErrorResumeNext)(function () {
    return new URL(href).protocol;
  });
  if (linkDefinition) {
    ariaLabelSegments.push(linkDefinition.title || (0, _botframeworkWebchatCore.onErrorResumeNext)(function () {
      return new URL(linkDefinition.url).host;
    }) || linkDefinition.url);

    // linkDefinition.identifier is uppercase, while linkDefinition.label is as-is.
    linkDefinition.label === textContent && classes.add('webchat__render-markdown__pure-identifier');
  }

  // For links that would be sanitized out, let's turn them into a button so we could handle them later.
  if (!allowedSchemes.map(function (scheme) {
    return "".concat(scheme, ":");
  }).includes(protocol)) {
    var _decoration$asButton;
    (_decoration$asButton = decoration.asButton) !== null && _decoration$asButton !== void 0 ? _decoration$asButton : decoration.asButton = true;
    classes.add('webchat__render-markdown__citation');
  } else if (protocol === 'http:' || protocol === 'https:') {
    decoration.iconClassName = [decoration.iconClassName, 'webchat__render-markdown__external-link-icon'].filter(function (className) {
      return className;
    }).join(' ');
    ariaLabelSegments.push(externalLinkAlt);
  }

  // The first segment is textContent. Putting textContent is aria-label is useless.
  if (ariaLabelSegments.length > 1) {
    var _decoration$ariaLabel;
    // If "aria-label" is already applied, do not overwrite it.
    (_decoration$ariaLabel = decoration.ariaLabel) !== null && _decoration$ariaLabel !== void 0 ? _decoration$ariaLabel : decoration.ariaLabel = function (value) {
      return value || ariaLabelSegments.join(' ');
    };
  }
  if (typeof (linkOptions === null || linkOptions === void 0 ? void 0 : linkOptions.className) === 'string') {
    classes.add(linkOptions.className);
  }

  // Resolve className
  var classNamesString = Array.from(classes).join(' ');
  if (linkOptions !== null && linkOptions !== void 0 && linkOptions.className && (linkOptions === null || linkOptions === void 0 ? void 0 : linkOptions.className) instanceof Function) {
    decoration.className = linkOptions.className(classNamesString);
  } else {
    decoration.className = classNamesString;
  }

  // By default, Markdown-It will set "title" to the link title in link definition.

  // However, "title" may be narrated by screen reader:
  // - Edge
  //   - <a> will narrate "aria-label" but not "title"
  //   - <button> will narrate both "aria-label" and "title"
  // - NVDA
  //   - <a> will narrate both "aria-label" and "title"
  //   - <button> will narrate both "aria-label" and "title"

  // Title makes it very difficult to control narrations by the screen reader. Thus, we are disabling it in favor of "aria-label".
  // This will not affect our accessibility compliance but UX. We could use a non-native tooltip or other forms of visual hint.

  (_decoration$title = decoration.title) !== null && _decoration$title !== void 0 ? _decoration$title : decoration.title = false;
  return decoration;
};
var useLocalizer = _botframeworkWebchatApi.hooks.useLocalizer;
var markdownIt = new _markdownIt.default().use(_betterLinks.default);
function isPlural(props) {
  return typeof props.stringIds !== 'string';
}
var LocalizedString = function LocalizedString(props) {
  var className = props.className,
    linkClassName = props.linkClassName,
    _props$onDecorateLink = props.onDecorateLink,
    onDecorateLink = _props$onDecorateLink === void 0 ? defaultDecorateLink : _props$onDecorateLink,
    stringIds = props.stringIds,
    values = props.values;
  var localize = useLocalizer(isPlural(props) && {
    plural: true
  });
  var env = (0, _react.useMemo)(function () {
    return {
      linkOptions: {
        className: linkClassName
      },
      decorateLink: onDecorateLink
    };
  }, [linkClassName, onDecorateLink]);
  var html = (0, _react.useMemo)(function () {
    return {
      __html: markdownIt.renderer.render(markdownIt.parseInline(localize.apply(void 0, [stringIds].concat(_toConsumableArray(values !== null && values !== void 0 ? values : []))), env), env)
    };
  }, [env, localize, stringIds, values]);
  return /*#__PURE__*/_react.default.createElement("span", {
    className: className,
    dangerouslySetInnerHTML: html
  });
};
var _default = /*#__PURE__*/(0, _react.memo)(LocalizedString);
exports.default = _default;
