// botframework-webchat-core/lib/types/external/OrgSchema/CreativeWork.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.parseCreativeWork = exports.creativeWork = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _valibot = require("valibot");
var _DefinedTerm = require("./DefinedTerm");
var _orgSchemaProperties = _interopRequireDefault(require("./private/orgSchemaProperties"));
var _orgSchemaProperty = _interopRequireDefault(require("./private/orgSchemaProperty"));
var _Thing = require("./Thing");
function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }
function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { (0, _defineProperty2["default"])(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
/**
 * The most generic kind of creative work, including books, movies, photographs, software programs, etc.
 *
 * This is partial implementation of https://schema.org/CreativeWork.
 *
 * @see https://schema.org/CreativeWork
 */
// Due to limitations of TypeScript, when using valibot.lazy(), the output type must be explicitly set.

var creativeWork = function creativeWork(entries) {
  return (0, _Thing.thing)(_objectSpread({
    // For forward compatibility, we did not enforce @type must be "CreativeWork" or any other subtypes.
    // In future, if Schema.org introduced a new subtype of CreativeWork, we should still able to parse that one as a CreativeWork.

    "abstract": (0, _orgSchemaProperty["default"])((0, _valibot.string)()),
    author: (0, _orgSchemaProperties["default"])((0, _valibot.string)()),
    citation: (0, _orgSchemaProperties["default"])((0, _valibot.lazy)(function () {
      return creativeWork();
    })),
    keywords: (0, _orgSchemaProperties["default"])((0, _valibot.union)([(0, _valibot.lazy)(function () {
      return (0, _DefinedTerm.definedTerm)();
    }), (0, _valibot.string)()])),
    pattern: (0, _orgSchemaProperty["default"])((0, _valibot.lazy)(function () {
      return (0, _DefinedTerm.definedTerm)();
    })),
    text: (0, _orgSchemaProperty["default"])((0, _valibot.string)()),
    usageInfo: (0, _orgSchemaProperty["default"])((0, _valibot.lazy)(function () {
      return creativeWork();
    }))
  }, entries));
};
exports.creativeWork = creativeWork;
var parseCreativeWork = function parseCreativeWork(data) {
  return (0, _valibot.parse)(creativeWork(), data);
};
exports.parseCreativeWork = parseCreativeWork;
