// botframework-webchat-core/lib/types/external/OrgSchema/Thing.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.thing = exports.parseThing = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _valibot = require("valibot");
var _Action = require("./Action");
var _orgSchemaProperties = _interopRequireDefault(require("./private/orgSchemaProperties"));
var _orgSchemaProperty = _interopRequireDefault(require("./private/orgSchemaProperty"));
function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }
function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { (0, _defineProperty2["default"])(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
/**
 * The most generic type of item.
 *
 * This is partial implementation of https://schema.org/Thing.
 *
 * @see https://schema.org/Thing
 */

var thingEntries = {
  '@context': (0, _valibot.optional)((0, _valibot.string)([(0, _valibot.value)('https://schema.org')])),
  '@id': (0, _valibot.optional)((0, _valibot.string)()),
  '@type': (0, _valibot.string)(),
  additionalType: (0, _orgSchemaProperty["default"])((0, _valibot.string)()),
  alternateName: (0, _orgSchemaProperty["default"])((0, _valibot.string)()),
  description: (0, _orgSchemaProperty["default"])((0, _valibot.string)()),
  name: (0, _orgSchemaProperty["default"])((0, _valibot.string)()),
  potentialAction: (0, _orgSchemaProperties["default"])((0, _valibot.lazy)(function () {
    return (0, _Action.action)();
  })),
  url: (0, _orgSchemaProperty["default"])((0, _valibot.string)())
};
var thing = function thing(entries) {
  return (0, _valibot.object)(_objectSpread(_objectSpread({}, thingEntries), entries),
  // Forward compatibility is the reason why we use unknown() here and not adhere to JSON-LD which drop unknown fields.
  //
  // Example:
  // - CreativeWork.editor must be type of Person (or any of its subtypes, Patient)
  // - Without unknown(), when we parse the CreativeWork, we will drop Patient.diagnosis
  // - That means, CreativeWork.editor.diagnosis will be unset despite CreativeWork.editor is of type Patient
  //
  // We can drop unknown() if there is a way to keep CreativeWork.editor.diagnosis.
  // It is okay to drop future/unsupported properties. But not today/supported properties.
  (0, _valibot.unknown)());
};
exports.thing = thing;
var parseThing = function parseThing(data) {
  return (0, _valibot.parse)(thing(), data);
};
exports.parseThing = parseThing;
