// react-chain-of-responsibility/dist/react-chain-of-responsibility.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/index.ts
var src_exports = {};
__export(src_exports, {
  createChainOfResponsibility: () => createChainOfResponsibility
});
module.exports = __toCommonJS(src_exports);

// src/createChainOfResponsibility.tsx
var import_prop_types = __toESM(require("prop-types"));
var import_react = __toESM(require("react"));

// src/isReactComponent.ts
function isConsumer(component) {
  var _a;
  return ((_a = component == null ? void 0 : component.$$typeof) == null ? void 0 : _a.toString()) === "Symbol(react.context)";
}
function isProvider(component) {
  var _a;
  return ((_a = component == null ? void 0 : component.$$typeof) == null ? void 0 : _a.toString()) === "Symbol(react.provider)";
}
function isFragment(component) {
  return (component == null ? void 0 : component.toString()) === "Symbol(react.fragment)";
}
function isFunctionComponent(component) {
  if (typeof component === "function") {
    return true;
  }
  return isPureFunctionComponent(component);
}
function isPureFunctionComponent(component) {
  var _a;
  return ((_a = component == null ? void 0 : component.$$typeof) == null ? void 0 : _a.toString()) === "Symbol(react.memo)" && isFunctionComponent(component.type);
}
function isComponentClass(component) {
  return typeof component === "object" && typeof (component == null ? void 0 : component["render"]) === "function";
}
function isReactComponent(component) {
  return isFunctionComponent(component) || isComponentClass(component) || isFragment(component) || isConsumer(component) || isProvider(component);
}

// src/private/compose.ts
function compose(...fns) {
  return (fn) => fns.reduce((chain, fn2) => fn2(chain), fn);
}

// src/private/applyMiddleware.ts
function applyMiddleware(...arrayOfMiddleware) {
  return (...init) => {
    const chain = arrayOfMiddleware.map((middleware) => middleware(...init));
    return compose(...chain);
  };
}

// src/createChainOfResponsibility.tsx
function createChainOfResponsibility(options = {}) {
  const defaultUseBuildComponentCallback = {
    get enhancer() {
      return void 0;
    },
    useBuildComponentCallback(_, options2) {
      if (options2 == null ? void 0 : options2.fallbackComponent) {
        return options2.fallbackComponent;
      }
      throw new Error("This component/hook cannot be used outside of its corresponding <Provider>");
    }
  };
  const context = (0, import_react.createContext)(defaultUseBuildComponentCallback);
  const Provider = ({ children, init, middleware }) => {
    if (!Array.isArray(middleware) || middleware.some((middleware2) => typeof middleware2 !== "function")) {
      throw new Error("middleware prop must be an array of functions");
    }
    const patchedMiddleware = Object.freeze(
      middleware ? middleware.map((fn) => (init2) => {
        const enhancer2 = fn(init2);
        return (next) => (originalRequest) => {
          let hasReturned;
          const returnValue = enhancer2((nextRequest) => {
            if (hasReturned) {
              throw new Error("next() cannot be called after the function had returned synchronously");
            }
            !options.passModifiedRequest && nextRequest !== originalRequest && console.warn(
              'react-chain-of-responsibility: "options.passModifiedRequest" must be set to true to pass a different request object to next().'
            );
            return next(options.passModifiedRequest ? nextRequest : originalRequest);
          })(originalRequest);
          hasReturned = true;
          if ((0, import_react.isValidElement)(returnValue)) {
            throw new Error("middleware must not return React element directly");
          } else if (returnValue !== false && returnValue !== null && typeof returnValue !== "undefined" && !isReactComponent(returnValue)) {
            throw new Error(
              "middleware must return false, null, undefined, function component, or class component"
            );
          }
          return returnValue;
        };
      }) : []
    );
    const { enhancer: parentEnhancer } = (0, import_react.useContext)(context);
    const enhancer = (0, import_react.useMemo)(
      () => (
        // We are reversing because it is easier to read:
        // - With reverse, [a, b, c] will become a(b(c(fn)))
        // - Without reverse, [a, b, c] will become c(b(a(fn)))
        applyMiddleware(
          ...[...patchedMiddleware, ...parentEnhancer ? [() => parentEnhancer] : []].reverse()
        )(init)
      ),
      [init, middleware, parentEnhancer]
    );
    const useBuildComponentCallback2 = (0, import_react.useCallback)(
      (request, options2 = {}) => enhancer(() => options2.fallbackComponent)(request),
      [enhancer]
    );
    const contextValue = (0, import_react.useMemo)(
      () => ({ enhancer, useBuildComponentCallback: useBuildComponentCallback2 }),
      [enhancer, useBuildComponentCallback2]
    );
    return /* @__PURE__ */ import_react.default.createElement(context.Provider, { value: contextValue }, children);
  };
  Provider.displayName = "ChainOfResponsibilityProvider";
  Provider.propTypes = {
    children: import_prop_types.default.any,
    init: import_prop_types.default.any,
    middleware: import_prop_types.default.any
  };
  const useBuildComponentCallback = () => (0, import_react.useContext)(context).useBuildComponentCallback;
  const Proxy2 = (0, import_react.memo)(
    // False positive: "children" is not a prop.
    // eslint-disable-next-line react/prop-types
    ({ children, fallbackComponent, request, ...props }) => {
      const enhancer = useBuildComponentCallback();
      const Component = enhancer(request, { fallbackComponent });
      return Component ? /* @__PURE__ */ import_react.default.createElement(Component, { ...props }, children) : null;
    }
  );
  Proxy2.displayName = "Proxy";
  Proxy2.propTypes = {
    fallbackComponent: import_prop_types.default.any,
    request: import_prop_types.default.any
  };
  return {
    Provider,
    Proxy: Proxy2,
    types: {
      init: void 0,
      middleware: void 0,
      props: void 0,
      request: void 0
    },
    useBuildComponentCallback
  };
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  createChainOfResponsibility
});
